{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential expression analysis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You have run the nf-core/rnaseq pipeline and checked the first quality control metrics of your fastq files. This was, however, only the primary analysis and we want to take it further.\n",
    "\n",
    "Due to the computational demand of the pipeline, you only ran the pipeline on two of the 16 samples in the study yesterday. We provide you an essential output of nf-core/rnaseq pipeline in the `data` folder: It contains the combined epression matrix as produced by Salmon, which provides transcript levels for each gene (rows) and each sample (columns).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We would now like to understand exactly the difference between the expression in our groups of mice. \n",
    "Which pipeline would you use for this?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "https://nf-co.re/differentialabundance/1.5.0/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Have a close look at the pipeline's \"Usage\" page on the [nf-core docs](nf-co.re). You will need to create a samplesheet (based on the column names in the provided matrix)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Please paste here the command you used. You may need to inspect the provided expression matrix more closely and create additional files, like a samplesheet (based on the column names) or a contrast file (there happens to also be one in `data/` that you can use)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--transcript-length-matrix 'salmon.merged.gene_lengths.tsv'\n",
    "--gtf ref-annotation/Mus_musculus.GRCm38.102.gtf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m\u001b[38;5;232m\u001b[48;5;43m N E X T F L O W \u001b[0;2m  ~  \u001b[mversion 25.04.7\u001b[m\n",
      "\u001b[K\n",
      "Launching\u001b[35m `https://github.com/nf-core/differentialabundance` \u001b[0;2m[\u001b[0;1;36msharp_hoover\u001b[0;2m] DSL2 - \u001b[36mrevision: \u001b[0;36m3dd360fed0 [1.5.0]\u001b[m\n",
      "\u001b[K\n",
      "\u001b[33mWARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\u001b[39m\u001b[K\n",
      "\n",
      "\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "                                        \u001b[0;32m,--.\u001b[0;30m/\u001b[0;32m,-.\u001b[0m\n",
      "\u001b[0;34m        ___     __   __   __   ___     \u001b[0;32m/,-._.--~'\u001b[0m\n",
      "\u001b[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \u001b[0;33m}  {\u001b[0m\n",
      "\u001b[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \u001b[0;32m\\`-._,-`-,\u001b[0m\n",
      "                                        \u001b[0;32m`._,._,'\u001b[0m\n",
      "\u001b[0;35m  nf-core/differentialabundance v1.5.0-g3dd360f\u001b[0m\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "\u001b[1mCore Nextflow options\u001b[0m\n",
      "  \u001b[0;34mrevision                    : \u001b[0;32m1.5.0\u001b[0m\n",
      "  \u001b[0;34mrunName                     : \u001b[0;32msharp_hoover\u001b[0m\n",
      "  \u001b[0;34mcontainerEngine             : \u001b[0;32mdocker\u001b[0m\n",
      "  \u001b[0;34mcontainer                   : \u001b[0;32m[RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\u001b[0m\n",
      "  \u001b[0;34mlaunchDir                   : \u001b[0;32m/home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_03\u001b[0m\n",
      "  \u001b[0;34mworkDir                     : \u001b[0;32m/home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_03/work\u001b[0m\n",
      "  \u001b[0;34mprojectDir                  : \u001b[0;32m/home/marcelullose/.nextflow/assets/nf-core/differentialabundance\u001b[0m\n",
      "  \u001b[0;34muserName                    : \u001b[0;32mmarcelullose\u001b[0m\n",
      "  \u001b[0;34mprofile                     : \u001b[0;32mrnaseq,docker\u001b[0m\n",
      "  \u001b[0;34mconfigFiles                 : \u001b[0;32m\u001b[0m\n",
      "\n",
      "\u001b[1mInput/output options\u001b[0m\n",
      "  \u001b[0;34minput                       : \u001b[0;32msamplesheet_diffabundance.csv\u001b[0m\n",
      "  \u001b[0;34mcontrasts                   : \u001b[0;32m./data/contrasts_venn.csv\u001b[0m\n",
      "  \u001b[0;34moutdir                      : \u001b[0;32mdifferentialabundance-out\u001b[0m\n",
      "\n",
      "\u001b[1mAbundance values\u001b[0m\n",
      "  \u001b[0;34mmatrix                      : \u001b[0;32mdata/salmon.merged.gene_counts.tsv\u001b[0m\n",
      "  \u001b[0;34maffy_cel_files_archive      : \u001b[0;32mnull\u001b[0m\n",
      "  \u001b[0;34mquerygse                    : \u001b[0;32mnull\u001b[0m\n",
      "\n",
      "\u001b[1mAffy input options\u001b[0m\n",
      "  \u001b[0;34maffy_cdfname                : \u001b[0;32mnull\u001b[0m\n",
      "\n",
      "\u001b[1mExploratory analysis\u001b[0m\n",
      "  \u001b[0;34mexploratory_log2_assays     : \u001b[0;32mraw,normalised\u001b[0m\n",
      "\n",
      "\u001b[1mDifferential analysis\u001b[0m\n",
      "  \u001b[0;34mdifferential_file_suffix    : \u001b[0;32m.deseq2.results.tsv\u001b[0m\n",
      "  \u001b[0;34mdifferential_min_fold_change: \u001b[0;32m0.5\u001b[0m\n",
      "\n",
      "\u001b[1mLimma specific options (microarray only)\u001b[0m\n",
      "  \u001b[0;34mlimma_spacing               : \u001b[0;32mnull\u001b[0m\n",
      "  \u001b[0;34mlimma_block                 : \u001b[0;32mnull\u001b[0m\n",
      "  \u001b[0;34mlimma_correlation           : \u001b[0;32mnull\u001b[0m\n",
      "\n",
      "\u001b[1mgprofiler2\u001b[0m\n",
      "  \u001b[0;34mgprofiler2_correction_method: \u001b[0;32mgSCS\u001b[0m\n",
      "  \u001b[0;34mgprofiler2_background_file  : \u001b[0;32mauto\u001b[0m\n",
      "\n",
      "\u001b[1mShiny app settings\u001b[0m\n",
      "  \u001b[0;34mshinyngs_shinyapps_account  : \u001b[0;32mnull\u001b[0m\n",
      "  \u001b[0;34mshinyngs_shinyapps_app_name : \u001b[0;32mnull\u001b[0m\n",
      "\n",
      "\u001b[1mOptions related to gene set analysis\u001b[0m\n",
      "  \u001b[0;34mgene_sets_files             : \u001b[0;32mnull\u001b[0m\n",
      "\n",
      "\u001b[1mReporting options\u001b[0m\n",
      "  \u001b[0;34mreport_title                : \u001b[0;32mnull\u001b[0m\n",
      "  \u001b[0;34mreport_author               : \u001b[0;32mnull\u001b[0m\n",
      "  \u001b[0;34mreport_description          : \u001b[0;32mnull\u001b[0m\n",
      "\n",
      "\u001b[1mInstitutional config options\u001b[0m\n",
      "  \u001b[0;34mconfig_profile_name         : \u001b[0;32mRNA-seq profile\u001b[0m\n",
      "  \u001b[0;34mconfig_profile_description  : \u001b[0;32mSettings for RNA-seq analysis\u001b[0m\n",
      "\n",
      "!! Only displaying parameters that differ from the pipeline defaults !!\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "If you use nf-core/differentialabundance for your analysis please cite:\n",
      "\n",
      "* The pipeline\n",
      "  https://doi.org/10.5281/zenodo.7568000\n",
      "\n",
      "* The nf-core framework\n",
      "  https://doi.org/10.1038/s41587-020-0439-x\n",
      "\n",
      "* Software dependencies\n",
      "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "\u001b[33mWARN: Access to undefined parameter `gtf` -- Initialise it to a default value eg. `params.gtf = some_value`\u001b[39m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…RENTIALABUNDANCE:\u001b[mVALIDATOR -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mCUSTOM_MATRIXFILTER -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mDESEQ2_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[6A\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…RENTIALABUNDANCE:\u001b[mVALIDATOR -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mCUSTOM_MATRIXFILTER -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mDESEQ2_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[8A\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…RENTIALABUNDANCE:\u001b[mVALIDATOR -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mCUSTOM_MATRIXFILTER -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mDESEQ2_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "\u001b[11A\n",
      "\u001b[2mexecutor >  local (1)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 0 of 1\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mCUSTOM_MATRIXFILTER -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mDESEQ2_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "\u001b[12A\n",
      "\u001b[2mexecutor >  local (1)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mCUSTOM_MATRIXFILTER -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mDESEQ2_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "\u001b[12A\n",
      "\u001b[2mexecutor >  local (2)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m8c/aaa19e\u001b[0;2m] \u001b[0;2m\u001b[mNFC…_MATRIXFILTER\u001b[33;2m (\u001b[0;33m[id:study]\u001b[2m)\u001b[m\u001b[2m |\u001b[m 0 of 1\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mDESEQ2_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "\u001b[12A\n",
      "\u001b[2mexecutor >  local (2)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m8c/aaa19e\u001b[0;2m] \u001b[0;2m\u001b[mNFC…_MATRIXFILTER\u001b[33;2m (\u001b[0;33m[id:study]\u001b[2m)\u001b[m\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NDANCE:\u001b[mDESEQ2_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "\u001b[12A\n",
      "\u001b[2mexecutor >  local (3)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m8c/aaa19e\u001b[0;2m] \u001b[0;2m\u001b[mNFC…_MATRIXFILTER\u001b[33;2m (\u001b[0;33m[id:study]\u001b[2m)\u001b[m\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM\u001b[2m |\u001b[m 0 of 1\u001b[K\n",
      "\u001b[2m[\u001b[0;34ma4/194845\u001b[0;2m] \u001b[0;2mNFC…rget:Sham_Sal, blocking:\u001b[m])\u001b[2m |\u001b[m 0 of 3\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "\u001b[12A\n",
      "\u001b[2mexecutor >  local (4)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m8c/aaa19e\u001b[0;2m] \u001b[0;2m\u001b[mNFC…_MATRIXFILTER\u001b[33;2m (\u001b[0;33m[id:study]\u001b[2m)\u001b[m\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM\u001b[2m |\u001b[m 0 of 1\u001b[K\n",
      "\u001b[2m[\u001b[0;34m33/18b350\u001b[0;2m] \u001b[0;2mNFC…rget:Sham_Sal, blocking:\u001b[m])\u001b[2m |\u001b[m 0 of 3\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "\u001b[31mERROR ~ Error executing process > 'NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])'\u001b[K\n",
      "\u001b[K\n",
      "Caused by:\u001b[K\n",
      "  Process `NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])` terminated with an error exit status (125)\u001b[K\n",
      "\u001b[K\n",
      "\u001b[K\n",
      "Command executed [/home/marcelullose/.nextflow/assets/nf-core/differentialabundance/modules/nf-core/deseq2/differential/templates/deseq_de.R]:\u001b[K\n",
      "\u001b[K\n",
      "  #!/usr/bin/env Rscript\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Functions                                  ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  #' Check for Non-Empty, Non-Whitespace String\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' This function checks if the input is non-NULL and contains more than just whitespace.\u001b[K\n",
      "  #' It returns TRUE if the input is a non-empty, non-whitespace string, and FALSE otherwise.\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param input A variable to check.\u001b[K\n",
      "  #' @return A logical value: TRUE if the input is a valid, non-empty, non-whitespace string; FALSE otherwise.\u001b[K\n",
      "  #' @examples\u001b[K\n",
      "  #' is_valid_string(\"Hello World\") # Returns TRUE\u001b[K\n",
      "  #' is_valid_string(\"   \")         # Returns FALSE\u001b[K\n",
      "  #' is_valid_string(NULL)          # Returns FALSE\u001b[K\n",
      "  \u001b[K\n",
      "  is_valid_string <- function(input) {\u001b[K\n",
      "      !is.null(input) && nzchar(trimws(input))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Parse out options from a string without recourse to optparse\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param x Long-form argument list like --opt1 val1 --opt2 val2\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return named list of options and values similar to optparse\u001b[K\n",
      "  \u001b[K\n",
      "  parse_args <- function(x){\u001b[K\n",
      "      args_list <- unlist(strsplit(x, ' ?--')[[1]])[-1]\u001b[K\n",
      "      args_vals <- lapply(args_list, function(x) scan(text=x, what='character', quiet = TRUE))\u001b[K\n",
      "  \u001b[K\n",
      "      # Ensure the option vectors are length 2 (key/ value) to catch empty ones\u001b[K\n",
      "      args_vals <- lapply(args_vals, function(z){ length(z) <- 2; z})\u001b[K\n",
      "  \u001b[K\n",
      "      parsed_args <- structure(lapply(args_vals, function(x) x[2]), names = lapply(args_vals, function(x) x[1]))\u001b[K\n",
      "      parsed_args[! is.na(parsed_args)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Flexibly read CSV or TSV files\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param file Input file\u001b[K\n",
      "  #' @param header Passed to read.delim()\u001b[K\n",
      "  #' @param row.names Passed to read.delim()\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  read_delim_flexible <- function(file, header = TRUE, row.names = NULL, check.names = TRUE){\u001b[K\n",
      "  \u001b[K\n",
      "      ext <- tolower(tail(strsplit(basename(file), split = \"\\\\.\")[[1]], 1))\u001b[K\n",
      "  \u001b[K\n",
      "      if (ext == \"tsv\" || ext == \"txt\") {\u001b[K\n",
      "          separator <- \"\\t\"\u001b[K\n",
      "      } else if (ext == \"csv\") {\u001b[K\n",
      "          separator <- \",\"\u001b[K\n",
      "      } else {\u001b[K\n",
      "          stop(paste(\"Unknown separator for\", ext))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      read.delim(\u001b[K\n",
      "          file,\u001b[K\n",
      "          sep = separator,\u001b[K\n",
      "          header = header,\u001b[K\n",
      "          row.names = row.names,\u001b[K\n",
      "          check.names = check.names\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Round numeric dataframe columns to fixed decimal places by applying\u001b[K\n",
      "  #' formatting and converting back to numerics\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param dataframe A data frame\u001b[K\n",
      "  #' @param columns Which columns to round (assumes all of them by default)\u001b[K\n",
      "  #' @param digits How many decimal places to round to?\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  round_dataframe_columns <- function(df, columns = NULL, digits = 8){\u001b[K\n",
      "      if (is.null(columns)){\u001b[K\n",
      "          columns <- colnames(df)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      df[,columns] <- format(\u001b[K\n",
      "          data.frame(df[, columns], check.names = FALSE),\u001b[K\n",
      "          nsmall = digits\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "      # Convert columns back to numeric\u001b[K\n",
      "  \u001b[K\n",
      "      for (c in columns) {\u001b[K\n",
      "          df[[c]][grep(\"^ *NA$\", df[[c]])] <- NA\u001b[K\n",
      "          df[[c]] <- as.numeric(df[[c]])\u001b[K\n",
      "      }\u001b[K\n",
      "      df\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## PARSE PARAMETERS FROM NEXTFLOW             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  # I've defined these in a single array like this so that we could go back to an\u001b[K\n",
      "  # optparse-driven method in future with module bin/ directories, rather than\u001b[K\n",
      "  # the template\u001b[K\n",
      "  \u001b[K\n",
      "  # Set defaults and classes\u001b[K\n",
      "  \u001b[K\n",
      "  opt <- list(\u001b[K\n",
      "      output_prefix = ifelse('Sham_oxy_vs_Sham_Sal' == 'null', 'Sham_oxy_vs_Sham_Sal', 'Sham_oxy_vs_Sham_Sal'),\u001b[K\n",
      "      count_file = 'study.filtered.tsv',\u001b[K\n",
      "      sample_file = 'samplesheet_diffabundance.sample_metadata.tsv',\u001b[K\n",
      "      contrast_variable = 'Condition',\u001b[K\n",
      "      reference_level = 'Sham_oxy',\u001b[K\n",
      "      target_level = 'Sham_Sal',\u001b[K\n",
      "      blocking_variables = NULL,\u001b[K\n",
      "      control_genes_file = '',\u001b[K\n",
      "      transcript_lengths_file = '',\u001b[K\n",
      "      sizefactors_from_controls = FALSE,\u001b[K\n",
      "      gene_id_col = \"gene_id\",\u001b[K\n",
      "      sample_id_col = \"experiment_accession\",\u001b[K\n",
      "      subset_to_contrast_samples = FALSE,\u001b[K\n",
      "      exclude_samples_col = NULL,\u001b[K\n",
      "      exclude_samples_values = NULL,\u001b[K\n",
      "      test = \"Wald\",\u001b[K\n",
      "      fit_type = \"parametric\",\u001b[K\n",
      "      sf_type = 'ratio',\u001b[K\n",
      "      min_replicates_for_replace = 7,\u001b[K\n",
      "      use_t = FALSE,\u001b[K\n",
      "      lfc_threshold = 0,\u001b[K\n",
      "      alt_hypothesis = 'greaterAbs',\u001b[K\n",
      "      independent_filtering = TRUE,\u001b[K\n",
      "      p_adjust_method = 'BH',\u001b[K\n",
      "      alpha = 0.1,\u001b[K\n",
      "      minmu = 0.5,\u001b[K\n",
      "      vs_method = 'vst', # 'rlog', 'vst', or 'rlog,vst'\u001b[K\n",
      "      shrink_lfc = TRUE,\u001b[K\n",
      "      cores = 1,\u001b[K\n",
      "      vs_blind = TRUE,\u001b[K\n",
      "      vst_nsub = 1000\u001b[K\n",
      "  )\u001b[K\n",
      "  opt_types <- lapply(opt, class)\u001b[K\n",
      "  \u001b[K\n",
      "  # Apply parameter overrides\u001b[K\n",
      "  \u001b[K\n",
      "  args_opt <- parse_args('--gene_id_col \"gene_id\" --sample_id_col \"sample\" --test Wald --fit_type parametric --sizefactors_from_controls false --sf_type ratio --min_replicates_for_replace 7 --use_t false --lfc_threshold 0 --alt_hypothesis greaterAbs --independent_filtering true --p_adjust_method BH --alpha 0.1 --minmu 0.5 --vs_method vst --vs_blind true --vst_nsub 1000 --shrink_lfc true --cores 1 --subset_to_contrast_samples false --blocking_variables')\u001b[K\n",
      "  for ( ao in names(args_opt)){\u001b[K\n",
      "      if (! ao %in% names(opt)){\u001b[K\n",
      "          stop(paste(\"Invalid option:\", ao))\u001b[K\n",
      "      }else{\u001b[K\n",
      "  \u001b[K\n",
      "          # Preserve classes from defaults where possible\u001b[K\n",
      "          if (! is.null(opt[[ao]])){\u001b[K\n",
      "              args_opt[[ao]] <- as(args_opt[[ao]], opt_types[[ao]])\u001b[K\n",
      "          }\u001b[K\n",
      "          opt[[ao]] <- args_opt[[ao]]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check if required parameters have been provided\u001b[K\n",
      "  \u001b[K\n",
      "  required_opts <- c('contrast_variable', 'reference_level', 'target_level', 'output_prefix')\u001b[K\n",
      "  missing <- required_opts[!unlist(lapply(opt[required_opts], is_valid_string)) | !required_opts %in% names(opt)]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing) > 0){\u001b[K\n",
      "      stop(paste(\"Missing required options:\", paste(missing, collapse=', ')))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check file inputs are valid\u001b[K\n",
      "  \u001b[K\n",
      "  for (file_input in c('count_file', 'sample_file')){\u001b[K\n",
      "      if (! is_valid_string(opt[[file_input]])) {\u001b[K\n",
      "          stop(paste(\"Please provide\", file_input), call. = FALSE)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      if (! file.exists(opt[[file_input]])){\u001b[K\n",
      "          stop(paste0('Value of ', file_input, ': ', opt[[file_input]], ' is not a valid file'))\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Finish loading libraries                   ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  library(DESeq2)\u001b[K\n",
      "  library(BiocParallel)\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## READ IN COUNTS FILE AND SAMPLE METADATA    ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  count.table <-\u001b[K\n",
      "      read_delim_flexible(\u001b[K\n",
      "          file = opt$count_file,\u001b[K\n",
      "          header = TRUE,\u001b[K\n",
      "          row.names = opt$gene_id_col,\u001b[K\n",
      "          check.names = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  sample.sheet <- read_delim_flexible(file = opt$sample_file)\u001b[K\n",
      "  \u001b[K\n",
      "  # Deal with spaces that may be in sample column\u001b[K\n",
      "  opt$sample_id_col <- make.names(opt$sample_id_col)\u001b[K\n",
      "  \u001b[K\n",
      "  if (! opt$sample_id_col %in% colnames(sample.sheet)){\u001b[K\n",
      "      stop(paste0(\"Specified sample ID column '\", opt$sample_id_col, \"' is not in the sample sheet\"))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Sample sheet can have duplicate rows for multiple sequencing runs, so uniqify\u001b[K\n",
      "  # before assigning row names\u001b[K\n",
      "  \u001b[K\n",
      "  sample.sheet <- sample.sheet[! duplicated(sample.sheet[[opt$sample_id_col]]), ]\u001b[K\n",
      "  rownames(sample.sheet) <- sample.sheet[[opt$sample_id_col]]\u001b[K\n",
      "  \u001b[K\n",
      "  # Check that all samples specified in the input sheet are present in the counts\u001b[K\n",
      "  # table. Assuming they are, subset and sort the count table to match the sample\u001b[K\n",
      "  # sheet\u001b[K\n",
      "  \u001b[K\n",
      "  missing_samples <-\u001b[K\n",
      "      sample.sheet[!rownames(sample.sheet) %in% colnames(count.table), opt$sample_id_col]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing_samples) > 0) {\u001b[K\n",
      "      stop(paste(\u001b[K\n",
      "          length(missing_samples),\u001b[K\n",
      "          'specified samples missing from count table:',\u001b[K\n",
      "          paste(missing_samples, collapse = ',')\u001b[K\n",
      "      ))\u001b[K\n",
      "  } else{\u001b[K\n",
      "      # Save any non-count data, will gene metadata etc we might need later\u001b[K\n",
      "      noncount.table <-\u001b[K\n",
      "          count.table[, !colnames(count.table) %in% rownames(sample.sheet), drop = FALSE]\u001b[K\n",
      "      count.table <- count.table[, rownames(sample.sheet)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## CHECK CONTRAST SPECIFICATION               ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast_variable <- make.names(opt$contrast_variable)\u001b[K\n",
      "  blocking.vars <- c()\u001b[K\n",
      "  \u001b[K\n",
      "  if (!contrast_variable %in% colnames(sample.sheet)) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste0(\u001b[K\n",
      "          'Chosen contrast variable \\\"',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          '\\\" not in sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (any(!c(opt$reflevel, opt$treatlevel) %in% sample.sheet[[contrast_variable]])) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste(\u001b[K\n",
      "          'Please choose reference and treatment levels that are present in the',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          'column of the sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      blocking.vars = make.names(unlist(strsplit(opt$blocking_variables, split = ';')))\u001b[K\n",
      "      if (!all(blocking.vars %in% colnames(sample.sheet))) {\u001b[K\n",
      "          missing_block <- paste(blocking.vars[! blocking.vars %in% colnames(sample.sheet)], collapse = ',')\u001b[K\n",
      "          stop(\u001b[K\n",
      "              paste(\u001b[K\n",
      "                  'Blocking variables', missing_block,\u001b[K\n",
      "                  'do not correspond to sample sheet columns.'\u001b[K\n",
      "              )\u001b[K\n",
      "          )\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, subset to only the samples involved in the contrast\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$subset_to_contrast_samples){\u001b[K\n",
      "      sample_selector <- sample.sheet[[contrast_variable]] %in% c(opt$target_level, opt$reference_level)\u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, remove samples with specified values in a given field (probably\u001b[K\n",
      "  # don't use this as well as the above)\u001b[K\n",
      "  \u001b[K\n",
      "  if ((is_valid_string(opt$exclude_samples_col)) && (is_valid_string(opt$exclude_samples_values))){\u001b[K\n",
      "      exclude_values = unlist(strsplit(opt$exclude_samples_values, split = ';'))\u001b[K\n",
      "  \u001b[K\n",
      "      if (! opt$exclude_samples_col %in% colnames(sample.sheet)){\u001b[K\n",
      "          stop(paste(opt$exclude_samples_col, ' specified to subset samples is not a valid sample sheet column'))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      print(paste0('Excluding samples with values of ', opt$exclude_samples_values, ' in ', opt$exclude_samples_col))\u001b[K\n",
      "      sample_selector <- ! sample.sheet[[opt$exclude_samples_col]] %in% exclude_values\u001b[K\n",
      "  \u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Now specify the model. Use cell-means style so we can be explicit with the\u001b[K\n",
      "  # contrasts\u001b[K\n",
      "  \u001b[K\n",
      "  model <- '~ 0'\u001b[K\n",
      "  \u001b[K\n",
      "  if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      model <- paste(model, paste(blocking.vars, collapse = ' + '), sep=' + ')\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Make sure all the appropriate variables are factors\u001b[K\n",
      "  \u001b[K\n",
      "  for (v in c(blocking.vars, contrast_variable)) {\u001b[K\n",
      "      sample.sheet[[v]] <- as.factor(sample.sheet[[v]])\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Variable of interest goes last, see\u001b[K\n",
      "  # https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs\u001b[K\n",
      "  \u001b[K\n",
      "  model <- paste(model, contrast_variable, sep = ' + ')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Run DESeq2 processes                       ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != ''){\u001b[K\n",
      "      control_genes <- readLines(opt$control_genes_file)\u001b[K\n",
      "      if (! opt$sizefactors_from_controls){\u001b[K\n",
      "          count.table <- count.table[setdiff(rownames(count.table), control_genes),]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeqDataSetFromMatrix(\u001b[K\n",
      "      countData = round(count.table),\u001b[K\n",
      "      colData = sample.sheet,\u001b[K\n",
      "      design = as.formula(model)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Build in transcript lengths. Copying what tximport does here:\u001b[K\n",
      "  # https://github.com/thelovelab/DESeq2/blob/6947d5bc629015fb8ffb2453a91b71665a164483/R/AllClasses.R#L409\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      lengths <-\u001b[K\n",
      "          read_delim_flexible(\u001b[K\n",
      "              file = opt$transcript_lengths_file,\u001b[K\n",
      "              header = TRUE,\u001b[K\n",
      "              row.names = opt$gene_id_col,\u001b[K\n",
      "              check.names = FALSE\u001b[K\n",
      "          )\u001b[K\n",
      "      lengths <- lengths[rownames(count.table), colnames(count.table)]\u001b[K\n",
      "      dimnames(lengths) <- dimnames(dds)\u001b[K\n",
      "      assays(dds)[[\"avgTxLength\"]] <- lengths\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != '' && opt$sizefactors_from_controls){\u001b[K\n",
      "      print(paste('Estimating size factors using', length(control_genes), 'control genes'))\u001b[K\n",
      "      dds <- estimateSizeFactors(dds, controlGenes=rownames(count.table) %in% control_genes)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeq(\u001b[K\n",
      "      dds,\u001b[K\n",
      "      test = opt$test,\u001b[K\n",
      "      fitType = opt$fit_type,\u001b[K\n",
      "      minReplicatesForReplace = opt$min_replicates_for_replace,\u001b[K\n",
      "      useT = opt$use_t,\u001b[K\n",
      "      sfType = opt$sf_type,\u001b[K\n",
      "      parallel=TRUE, BPPARAM=MulticoreParam(opt$cores)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  comp.results <-\u001b[K\n",
      "      results(\u001b[K\n",
      "          dds,\u001b[K\n",
      "          lfcThreshold = opt$lfc_threshold,\u001b[K\n",
      "          altHypothesis = opt$alt_hypothesis,\u001b[K\n",
      "          independentFiltering = opt$independent_filtering,\u001b[K\n",
      "          alpha = opt$alpha,\u001b[K\n",
      "          pAdjustMethod = opt$p_adjust_method,\u001b[K\n",
      "          minmu = opt$minmu,\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$shrink_lfc){\u001b[K\n",
      "      comp.results <- lfcShrink(dds,\u001b[K\n",
      "          type = 'ashr',\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # See https://support.bioconductor.org/p/97676/\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      size_factors = estimateSizeFactorsForMatrix(counts(dds) / assays(dds)[[\"avgTxLength\"]])\u001b[K\n",
      "  }else {\u001b[K\n",
      "      size_factors = sizeFactors(dds)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Generate outputs                           ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast.name <-\u001b[K\n",
      "      paste(opt$target_level, opt$reference_level, sep = \"_vs_\")\u001b[K\n",
      "  cat(\"Saving results for \", contrast.name, \" ...\u001b[K\n",
      "  \", sep = \"\")\u001b[K\n",
      "  \u001b[K\n",
      "  # Differential expression table- note very limited rounding for consistency of\u001b[K\n",
      "  # results\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(comp.results)), opt$gene_id_col),\u001b[K\n",
      "      round_dataframe_columns(\u001b[K\n",
      "          data.frame(comp.results[, !(colnames(comp.results) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "      )\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.results.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Dispersion plot\u001b[K\n",
      "  \u001b[K\n",
      "  png(\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.dispersion.png', sep = '.'),\u001b[K\n",
      "      width = 600,\u001b[K\n",
      "      height = 600\u001b[K\n",
      "  )\u001b[K\n",
      "  plotDispEsts(dds)\u001b[K\n",
      "  dev.off()\u001b[K\n",
      "  \u001b[K\n",
      "  # R object for other processes to use\u001b[K\n",
      "  \u001b[K\n",
      "  saveRDS(dds, file = paste(opt$output_prefix, 'dds.rld.rds', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  # Size factors\u001b[K\n",
      "  \u001b[K\n",
      "  sf_df = data.frame(\u001b[K\n",
      "      sample = names(size_factors),\u001b[K\n",
      "      data.frame(size_factors, check.names = FALSE),\u001b[K\n",
      "      check.names = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  colnames(sf_df) <- c('sample', 'sizeFactor')\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      sf_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.sizefactors.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Write specified matrices\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "      data.frame(counts(dds, normalized = TRUE)[, !(colnames(counts(dds, normalized = TRUE)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'normalised_counts.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Note very limited rounding for consistency of results\u001b[K\n",
      "  \u001b[K\n",
      "  for (vs_method_name in strsplit(opt$vs_method, ',')){\u001b[K\n",
      "      if (vs_method_name == 'vst'){\u001b[K\n",
      "          vs_mat <- vst(dds, blind = opt$vs_blind, nsub = opt$vst_nsub)\u001b[K\n",
      "      }else if (vs_method_name == 'rlog'){\u001b[K\n",
      "          vs_mat <- rlog(dds, blind = opt$vs_blind, fitType = opt$fit_type)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      # Again apply the slight rounding and then restore numeric\u001b[K\n",
      "  \u001b[K\n",
      "      out_df <- cbind(\u001b[K\n",
      "          setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "          round_dataframe_columns(\u001b[K\n",
      "              data.frame(assay(vs_mat)[, !(colnames(assay(vs_mat)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "      write.table(\u001b[K\n",
      "          out_df,\u001b[K\n",
      "          file = paste(opt$output_prefix, vs_method_name, 'tsv', sep = '.'),\u001b[K\n",
      "          col.names = TRUE,\u001b[K\n",
      "          row.names = FALSE,\u001b[K\n",
      "          sep = '\t',\u001b[K\n",
      "          quote = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Save model to file\u001b[K\n",
      "  \u001b[K\n",
      "  write(model, file=paste(opt$output_prefix, 'deseq2.model.txt', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## R SESSION INFO                             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  sink(paste(opt$output_prefix, \"R_sessionInfo.log\", sep = '.'))\u001b[K\n",
      "  print(sessionInfo())\u001b[K\n",
      "  sink()\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## VERSIONS FILE                              ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  r.version <- strsplit(version[['version.string']], ' ')[[1]][3]\u001b[K\n",
      "  deseq2.version <- as.character(packageVersion('DESeq2'))\u001b[K\n",
      "  \u001b[K\n",
      "  writeLines(\u001b[K\n",
      "      c(\u001b[K\n",
      "          '\"NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL\":',\u001b[K\n",
      "          paste('    r-base:', r.version),\u001b[K\n",
      "          paste('    bioconductor-deseq2:', deseq2.version)\u001b[K\n",
      "      ),\u001b[K\n",
      "  'versions.yml')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "\u001b[K\n",
      "Command exit status:\u001b[K\n",
      "  125\u001b[K\n",
      "\u001b[K\n",
      "Command output:\u001b[K\n",
      "  (empty)\u001b[K\n",
      "\u001b[K\n",
      "Command error:\u001b[K\n",
      "  Loading required package: matrixStats\u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘MatrixGenerics’\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\u001b[K\n",
      "      colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\u001b[K\n",
      "      colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\u001b[K\n",
      "      colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\u001b[K\n",
      "      colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\u001b[K\n",
      "      colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\u001b[K\n",
      "      colWeightedMeans, colWeightedMedians, colWeightedSds,\u001b[K\n",
      "      colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\u001b[K\n",
      "      rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\u001b[K\n",
      "      rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\u001b[K\n",
      "      rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\u001b[K\n",
      "      rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\u001b[K\n",
      "      rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\u001b[K\n",
      "      rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\u001b[K\n",
      "      rowWeightedSds, rowWeightedVars\u001b[K\n",
      "  \u001b[K\n",
      "  Loading required package: Biobase\u001b[K\n",
      "  Welcome to Bioconductor\u001b[K\n",
      "  \u001b[K\n",
      "      Vignettes contain introductory material; view with\u001b[K\n",
      "      'browseVignettes()'. To cite Bioconductor, see\u001b[K\n",
      "      'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\u001b[K\n",
      "  \u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘Biobase’\u001b[K\n",
      "  \u001b[K\n",
      "  The following object is masked from ‘package:MatrixGenerics’:\u001b[K\n",
      "  \u001b[K\n",
      "      rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      anyMissing, rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  converting counts to integer mode\u001b[K\n",
      "  estimating size factors\u001b[K\n",
      "  estimating dispersions\u001b[K\n",
      "  gene-wise dispersion estimates: 1 workers\u001b[K\n",
      "  mean-dispersion relationship\u001b[K\n",
      "  final dispersion estimates, fitting model and testing: 1 workers\u001b[K\n",
      "  using 'ashr' for LFC shrinkage. If used in published research, please cite:\u001b[K\n",
      "      Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\u001b[K\n",
      "      https://doi.org/10.1093/biostatistics/kxw041\u001b[K\n",
      "  time=\"2025-10-01T11:42:37+02:00\" level=error msg=\"error waiting for container: unexpected EOF\"\u001b[K\n",
      "\u001b[K\n",
      "Work dir:\u001b[K\n",
      "  /home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_03/work/a4/194845d6f6d9a0fc47f929230de056\u001b[K\n",
      "\u001b[K\n",
      "Container:\u001b[K\n",
      "  quay.io/biocontainers/bioconductor-deseq2:1.34.0--r41hc247a5b_3\u001b[K\n",
      "\u001b[K\n",
      "Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`\u001b[K\n",
      "\u001b[K\n",
      " -- Check '.nextflow.log' file for details\u001b[39m\u001b[K\n",
      "\u001b[632A\n",
      "\u001b[2mexecutor >  local (4)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m8c/aaa19e\u001b[0;2m] \u001b[0;2m\u001b[mNFC…_MATRIXFILTER\u001b[33;2m (\u001b[0;33m[id:study]\u001b[2m)\u001b[m\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM\u001b[2m |\u001b[m 0 of 1\u001b[K\n",
      "\u001b[2m[\u001b[0;34ma4/194845\u001b[0;2m] \u001b[0;2mNFC…rget:Sham_Sal, blocking:\u001b[m])\u001b[2m |\u001b[m 0 of 3\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "Execution cancelled -- Finishing pending tasks before exit\u001b[K\n",
      "\u001b[31mERROR ~ Error executing process > 'NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])'\u001b[K\n",
      "\u001b[K\n",
      "Caused by:\u001b[K\n",
      "  Process `NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])` terminated with an error exit status (125)\u001b[K\n",
      "\u001b[K\n",
      "\u001b[K\n",
      "Command executed [/home/marcelullose/.nextflow/assets/nf-core/differentialabundance/modules/nf-core/deseq2/differential/templates/deseq_de.R]:\u001b[K\n",
      "\u001b[K\n",
      "  #!/usr/bin/env Rscript\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Functions                                  ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  #' Check for Non-Empty, Non-Whitespace String\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' This function checks if the input is non-NULL and contains more than just whitespace.\u001b[K\n",
      "  #' It returns TRUE if the input is a non-empty, non-whitespace string, and FALSE otherwise.\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param input A variable to check.\u001b[K\n",
      "  #' @return A logical value: TRUE if the input is a valid, non-empty, non-whitespace string; FALSE otherwise.\u001b[K\n",
      "  #' @examples\u001b[K\n",
      "  #' is_valid_string(\"Hello World\") # Returns TRUE\u001b[K\n",
      "  #' is_valid_string(\"   \")         # Returns FALSE\u001b[K\n",
      "  #' is_valid_string(NULL)          # Returns FALSE\u001b[K\n",
      "  \u001b[K\n",
      "  is_valid_string <- function(input) {\u001b[K\n",
      "      !is.null(input) && nzchar(trimws(input))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Parse out options from a string without recourse to optparse\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param x Long-form argument list like --opt1 val1 --opt2 val2\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return named list of options and values similar to optparse\u001b[K\n",
      "  \u001b[K\n",
      "  parse_args <- function(x){\u001b[K\n",
      "      args_list <- unlist(strsplit(x, ' ?--')[[1]])[-1]\u001b[K\n",
      "      args_vals <- lapply(args_list, function(x) scan(text=x, what='character', quiet = TRUE))\u001b[K\n",
      "  \u001b[K\n",
      "      # Ensure the option vectors are length 2 (key/ value) to catch empty ones\u001b[K\n",
      "      args_vals <- lapply(args_vals, function(z){ length(z) <- 2; z})\u001b[K\n",
      "  \u001b[K\n",
      "      parsed_args <- structure(lapply(args_vals, function(x) x[2]), names = lapply(args_vals, function(x) x[1]))\u001b[K\n",
      "      parsed_args[! is.na(parsed_args)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Flexibly read CSV or TSV files\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param file Input file\u001b[K\n",
      "  #' @param header Passed to read.delim()\u001b[K\n",
      "  #' @param row.names Passed to read.delim()\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  read_delim_flexible <- function(file, header = TRUE, row.names = NULL, check.names = TRUE){\u001b[K\n",
      "  \u001b[K\n",
      "      ext <- tolower(tail(strsplit(basename(file), split = \"\\\\.\")[[1]], 1))\u001b[K\n",
      "  \u001b[K\n",
      "      if (ext == \"tsv\" || ext == \"txt\") {\u001b[K\n",
      "          separator <- \"\\t\"\u001b[K\n",
      "      } else if (ext == \"csv\") {\u001b[K\n",
      "          separator <- \",\"\u001b[K\n",
      "      } else {\u001b[K\n",
      "          stop(paste(\"Unknown separator for\", ext))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      read.delim(\u001b[K\n",
      "          file,\u001b[K\n",
      "          sep = separator,\u001b[K\n",
      "          header = header,\u001b[K\n",
      "          row.names = row.names,\u001b[K\n",
      "          check.names = check.names\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Round numeric dataframe columns to fixed decimal places by applying\u001b[K\n",
      "  #' formatting and converting back to numerics\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param dataframe A data frame\u001b[K\n",
      "  #' @param columns Which columns to round (assumes all of them by default)\u001b[K\n",
      "  #' @param digits How many decimal places to round to?\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  round_dataframe_columns <- function(df, columns = NULL, digits = 8){\u001b[K\n",
      "      if (is.null(columns)){\u001b[K\n",
      "          columns <- colnames(df)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      df[,columns] <- format(\u001b[K\n",
      "          data.frame(df[, columns], check.names = FALSE),\u001b[K\n",
      "          nsmall = digits\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "      # Convert columns back to numeric\u001b[K\n",
      "  \u001b[K\n",
      "      for (c in columns) {\u001b[K\n",
      "          df[[c]][grep(\"^ *NA$\", df[[c]])] <- NA\u001b[K\n",
      "          df[[c]] <- as.numeric(df[[c]])\u001b[K\n",
      "      }\u001b[K\n",
      "      df\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## PARSE PARAMETERS FROM NEXTFLOW             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  # I've defined these in a single array like this so that we could go back to an\u001b[K\n",
      "  # optparse-driven method in future with module bin/ directories, rather than\u001b[K\n",
      "  # the template\u001b[K\n",
      "  \u001b[K\n",
      "  # Set defaults and classes\u001b[K\n",
      "  \u001b[K\n",
      "  opt <- list(\u001b[K\n",
      "      output_prefix = ifelse('Sham_oxy_vs_Sham_Sal' == 'null', 'Sham_oxy_vs_Sham_Sal', 'Sham_oxy_vs_Sham_Sal'),\u001b[K\n",
      "      count_file = 'study.filtered.tsv',\u001b[K\n",
      "      sample_file = 'samplesheet_diffabundance.sample_metadata.tsv',\u001b[K\n",
      "      contrast_variable = 'Condition',\u001b[K\n",
      "      reference_level = 'Sham_oxy',\u001b[K\n",
      "      target_level = 'Sham_Sal',\u001b[K\n",
      "      blocking_variables = NULL,\u001b[K\n",
      "      control_genes_file = '',\u001b[K\n",
      "      transcript_lengths_file = '',\u001b[K\n",
      "      sizefactors_from_controls = FALSE,\u001b[K\n",
      "      gene_id_col = \"gene_id\",\u001b[K\n",
      "      sample_id_col = \"experiment_accession\",\u001b[K\n",
      "      subset_to_contrast_samples = FALSE,\u001b[K\n",
      "      exclude_samples_col = NULL,\u001b[K\n",
      "      exclude_samples_values = NULL,\u001b[K\n",
      "      test = \"Wald\",\u001b[K\n",
      "      fit_type = \"parametric\",\u001b[K\n",
      "      sf_type = 'ratio',\u001b[K\n",
      "      min_replicates_for_replace = 7,\u001b[K\n",
      "      use_t = FALSE,\u001b[K\n",
      "      lfc_threshold = 0,\u001b[K\n",
      "      alt_hypothesis = 'greaterAbs',\u001b[K\n",
      "      independent_filtering = TRUE,\u001b[K\n",
      "      p_adjust_method = 'BH',\u001b[K\n",
      "      alpha = 0.1,\u001b[K\n",
      "      minmu = 0.5,\u001b[K\n",
      "      vs_method = 'vst', # 'rlog', 'vst', or 'rlog,vst'\u001b[K\n",
      "      shrink_lfc = TRUE,\u001b[K\n",
      "      cores = 1,\u001b[K\n",
      "      vs_blind = TRUE,\u001b[K\n",
      "      vst_nsub = 1000\u001b[K\n",
      "  )\u001b[K\n",
      "  opt_types <- lapply(opt, class)\u001b[K\n",
      "  \u001b[K\n",
      "  # Apply parameter overrides\u001b[K\n",
      "  \u001b[K\n",
      "  args_opt <- parse_args('--gene_id_col \"gene_id\" --sample_id_col \"sample\" --test Wald --fit_type parametric --sizefactors_from_controls false --sf_type ratio --min_replicates_for_replace 7 --use_t false --lfc_threshold 0 --alt_hypothesis greaterAbs --independent_filtering true --p_adjust_method BH --alpha 0.1 --minmu 0.5 --vs_method vst --vs_blind true --vst_nsub 1000 --shrink_lfc true --cores 1 --subset_to_contrast_samples false --blocking_variables')\u001b[K\n",
      "  for ( ao in names(args_opt)){\u001b[K\n",
      "      if (! ao %in% names(opt)){\u001b[K\n",
      "          stop(paste(\"Invalid option:\", ao))\u001b[K\n",
      "      }else{\u001b[K\n",
      "  \u001b[K\n",
      "          # Preserve classes from defaults where possible\u001b[K\n",
      "          if (! is.null(opt[[ao]])){\u001b[K\n",
      "              args_opt[[ao]] <- as(args_opt[[ao]], opt_types[[ao]])\u001b[K\n",
      "          }\u001b[K\n",
      "          opt[[ao]] <- args_opt[[ao]]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check if required parameters have been provided\u001b[K\n",
      "  \u001b[K\n",
      "  required_opts <- c('contrast_variable', 'reference_level', 'target_level', 'output_prefix')\u001b[K\n",
      "  missing <- required_opts[!unlist(lapply(opt[required_opts], is_valid_string)) | !required_opts %in% names(opt)]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing) > 0){\u001b[K\n",
      "      stop(paste(\"Missing required options:\", paste(missing, collapse=', ')))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check file inputs are valid\u001b[K\n",
      "  \u001b[K\n",
      "  for (file_input in c('count_file', 'sample_file')){\u001b[K\n",
      "      if (! is_valid_string(opt[[file_input]])) {\u001b[K\n",
      "          stop(paste(\"Please provide\", file_input), call. = FALSE)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      if (! file.exists(opt[[file_input]])){\u001b[K\n",
      "          stop(paste0('Value of ', file_input, ': ', opt[[file_input]], ' is not a valid file'))\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Finish loading libraries                   ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  library(DESeq2)\u001b[K\n",
      "  library(BiocParallel)\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## READ IN COUNTS FILE AND SAMPLE METADATA    ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  count.table <-\u001b[K\n",
      "      read_delim_flexible(\u001b[K\n",
      "          file = opt$count_file,\u001b[K\n",
      "          header = TRUE,\u001b[K\n",
      "          row.names = opt$gene_id_col,\u001b[K\n",
      "          check.names = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  sample.sheet <- read_delim_flexible(file = opt$sample_file)\u001b[K\n",
      "  \u001b[K\n",
      "  # Deal with spaces that may be in sample column\u001b[K\n",
      "  opt$sample_id_col <- make.names(opt$sample_id_col)\u001b[K\n",
      "  \u001b[K\n",
      "  if (! opt$sample_id_col %in% colnames(sample.sheet)){\u001b[K\n",
      "      stop(paste0(\"Specified sample ID column '\", opt$sample_id_col, \"' is not in the sample sheet\"))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Sample sheet can have duplicate rows for multiple sequencing runs, so uniqify\u001b[K\n",
      "  # before assigning row names\u001b[K\n",
      "  \u001b[K\n",
      "  sample.sheet <- sample.sheet[! duplicated(sample.sheet[[opt$sample_id_col]]), ]\u001b[K\n",
      "  rownames(sample.sheet) <- sample.sheet[[opt$sample_id_col]]\u001b[K\n",
      "  \u001b[K\n",
      "  # Check that all samples specified in the input sheet are present in the counts\u001b[K\n",
      "  # table. Assuming they are, subset and sort the count table to match the sample\u001b[K\n",
      "  # sheet\u001b[K\n",
      "  \u001b[K\n",
      "  missing_samples <-\u001b[K\n",
      "      sample.sheet[!rownames(sample.sheet) %in% colnames(count.table), opt$sample_id_col]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing_samples) > 0) {\u001b[K\n",
      "      stop(paste(\u001b[K\n",
      "          length(missing_samples),\u001b[K\n",
      "          'specified samples missing from count table:',\u001b[K\n",
      "          paste(missing_samples, collapse = ',')\u001b[K\n",
      "      ))\u001b[K\n",
      "  } else{\u001b[K\n",
      "      # Save any non-count data, will gene metadata etc we might need later\u001b[K\n",
      "      noncount.table <-\u001b[K\n",
      "          count.table[, !colnames(count.table) %in% rownames(sample.sheet), drop = FALSE]\u001b[K\n",
      "      count.table <- count.table[, rownames(sample.sheet)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## CHECK CONTRAST SPECIFICATION               ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast_variable <- make.names(opt$contrast_variable)\u001b[K\n",
      "  blocking.vars <- c()\u001b[K\n",
      "  \u001b[K\n",
      "  if (!contrast_variable %in% colnames(sample.sheet)) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste0(\u001b[K\n",
      "          'Chosen contrast variable \\\"',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          '\\\" not in sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (any(!c(opt$reflevel, opt$treatlevel) %in% sample.sheet[[contrast_variable]])) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste(\u001b[K\n",
      "          'Please choose reference and treatment levels that are present in the',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          'column of the sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      blocking.vars = make.names(unlist(strsplit(opt$blocking_variables, split = ';')))\u001b[K\n",
      "      if (!all(blocking.vars %in% colnames(sample.sheet))) {\u001b[K\n",
      "          missing_block <- paste(blocking.vars[! blocking.vars %in% colnames(sample.sheet)], collapse = ',')\u001b[K\n",
      "          stop(\u001b[K\n",
      "              paste(\u001b[K\n",
      "                  'Blocking variables', missing_block,\u001b[K\n",
      "                  'do not correspond to sample sheet columns.'\u001b[K\n",
      "              )\u001b[K\n",
      "          )\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, subset to only the samples involved in the contrast\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$subset_to_contrast_samples){\u001b[K\n",
      "      sample_selector <- sample.sheet[[contrast_variable]] %in% c(opt$target_level, opt$reference_level)\u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, remove samples with specified values in a given field (probably\u001b[K\n",
      "  # don't use this as well as the above)\u001b[K\n",
      "  \u001b[K\n",
      "  if ((is_valid_string(opt$exclude_samples_col)) && (is_valid_string(opt$exclude_samples_values))){\u001b[K\n",
      "      exclude_values = unlist(strsplit(opt$exclude_samples_values, split = ';'))\u001b[K\n",
      "  \u001b[K\n",
      "      if (! opt$exclude_samples_col %in% colnames(sample.sheet)){\u001b[K\n",
      "          stop(paste(opt$exclude_samples_col, ' specified to subset samples is not a valid sample sheet column'))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      print(paste0('Excluding samples with values of ', opt$exclude_samples_values, ' in ', opt$exclude_samples_col))\u001b[K\n",
      "      sample_selector <- ! sample.sheet[[opt$exclude_samples_col]] %in% exclude_values\u001b[K\n",
      "  \u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Now specify the model. Use cell-means style so we can be explicit with the\u001b[K\n",
      "  # contrasts\u001b[K\n",
      "  \u001b[K\n",
      "  model <- '~ 0'\u001b[K\n",
      "  \u001b[K\n",
      "  if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      model <- paste(model, paste(blocking.vars, collapse = ' + '), sep=' + ')\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Make sure all the appropriate variables are factors\u001b[K\n",
      "  \u001b[K\n",
      "  for (v in c(blocking.vars, contrast_variable)) {\u001b[K\n",
      "      sample.sheet[[v]] <- as.factor(sample.sheet[[v]])\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Variable of interest goes last, see\u001b[K\n",
      "  # https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs\u001b[K\n",
      "  \u001b[K\n",
      "  model <- paste(model, contrast_variable, sep = ' + ')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Run DESeq2 processes                       ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != ''){\u001b[K\n",
      "      control_genes <- readLines(opt$control_genes_file)\u001b[K\n",
      "      if (! opt$sizefactors_from_controls){\u001b[K\n",
      "          count.table <- count.table[setdiff(rownames(count.table), control_genes),]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeqDataSetFromMatrix(\u001b[K\n",
      "      countData = round(count.table),\u001b[K\n",
      "      colData = sample.sheet,\u001b[K\n",
      "      design = as.formula(model)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Build in transcript lengths. Copying what tximport does here:\u001b[K\n",
      "  # https://github.com/thelovelab/DESeq2/blob/6947d5bc629015fb8ffb2453a91b71665a164483/R/AllClasses.R#L409\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      lengths <-\u001b[K\n",
      "          read_delim_flexible(\u001b[K\n",
      "              file = opt$transcript_lengths_file,\u001b[K\n",
      "              header = TRUE,\u001b[K\n",
      "              row.names = opt$gene_id_col,\u001b[K\n",
      "              check.names = FALSE\u001b[K\n",
      "          )\u001b[K\n",
      "      lengths <- lengths[rownames(count.table), colnames(count.table)]\u001b[K\n",
      "      dimnames(lengths) <- dimnames(dds)\u001b[K\n",
      "      assays(dds)[[\"avgTxLength\"]] <- lengths\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != '' && opt$sizefactors_from_controls){\u001b[K\n",
      "      print(paste('Estimating size factors using', length(control_genes), 'control genes'))\u001b[K\n",
      "      dds <- estimateSizeFactors(dds, controlGenes=rownames(count.table) %in% control_genes)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeq(\u001b[K\n",
      "      dds,\u001b[K\n",
      "      test = opt$test,\u001b[K\n",
      "      fitType = opt$fit_type,\u001b[K\n",
      "      minReplicatesForReplace = opt$min_replicates_for_replace,\u001b[K\n",
      "      useT = opt$use_t,\u001b[K\n",
      "      sfType = opt$sf_type,\u001b[K\n",
      "      parallel=TRUE, BPPARAM=MulticoreParam(opt$cores)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  comp.results <-\u001b[K\n",
      "      results(\u001b[K\n",
      "          dds,\u001b[K\n",
      "          lfcThreshold = opt$lfc_threshold,\u001b[K\n",
      "          altHypothesis = opt$alt_hypothesis,\u001b[K\n",
      "          independentFiltering = opt$independent_filtering,\u001b[K\n",
      "          alpha = opt$alpha,\u001b[K\n",
      "          pAdjustMethod = opt$p_adjust_method,\u001b[K\n",
      "          minmu = opt$minmu,\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$shrink_lfc){\u001b[K\n",
      "      comp.results <- lfcShrink(dds,\u001b[K\n",
      "          type = 'ashr',\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # See https://support.bioconductor.org/p/97676/\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      size_factors = estimateSizeFactorsForMatrix(counts(dds) / assays(dds)[[\"avgTxLength\"]])\u001b[K\n",
      "  }else {\u001b[K\n",
      "      size_factors = sizeFactors(dds)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Generate outputs                           ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast.name <-\u001b[K\n",
      "      paste(opt$target_level, opt$reference_level, sep = \"_vs_\")\u001b[K\n",
      "  cat(\"Saving results for \", contrast.name, \" ...\u001b[K\n",
      "  \", sep = \"\")\u001b[K\n",
      "  \u001b[K\n",
      "  # Differential expression table- note very limited rounding for consistency of\u001b[K\n",
      "  # results\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(comp.results)), opt$gene_id_col),\u001b[K\n",
      "      round_dataframe_columns(\u001b[K\n",
      "          data.frame(comp.results[, !(colnames(comp.results) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "      )\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.results.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Dispersion plot\u001b[K\n",
      "  \u001b[K\n",
      "  png(\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.dispersion.png', sep = '.'),\u001b[K\n",
      "      width = 600,\u001b[K\n",
      "      height = 600\u001b[K\n",
      "  )\u001b[K\n",
      "  plotDispEsts(dds)\u001b[K\n",
      "  dev.off()\u001b[K\n",
      "  \u001b[K\n",
      "  # R object for other processes to use\u001b[K\n",
      "  \u001b[K\n",
      "  saveRDS(dds, file = paste(opt$output_prefix, 'dds.rld.rds', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  # Size factors\u001b[K\n",
      "  \u001b[K\n",
      "  sf_df = data.frame(\u001b[K\n",
      "      sample = names(size_factors),\u001b[K\n",
      "      data.frame(size_factors, check.names = FALSE),\u001b[K\n",
      "      check.names = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  colnames(sf_df) <- c('sample', 'sizeFactor')\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      sf_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.sizefactors.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Write specified matrices\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "      data.frame(counts(dds, normalized = TRUE)[, !(colnames(counts(dds, normalized = TRUE)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'normalised_counts.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Note very limited rounding for consistency of results\u001b[K\n",
      "  \u001b[K\n",
      "  for (vs_method_name in strsplit(opt$vs_method, ',')){\u001b[K\n",
      "      if (vs_method_name == 'vst'){\u001b[K\n",
      "          vs_mat <- vst(dds, blind = opt$vs_blind, nsub = opt$vst_nsub)\u001b[K\n",
      "      }else if (vs_method_name == 'rlog'){\u001b[K\n",
      "          vs_mat <- rlog(dds, blind = opt$vs_blind, fitType = opt$fit_type)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      # Again apply the slight rounding and then restore numeric\u001b[K\n",
      "  \u001b[K\n",
      "      out_df <- cbind(\u001b[K\n",
      "          setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "          round_dataframe_columns(\u001b[K\n",
      "              data.frame(assay(vs_mat)[, !(colnames(assay(vs_mat)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "      write.table(\u001b[K\n",
      "          out_df,\u001b[K\n",
      "          file = paste(opt$output_prefix, vs_method_name, 'tsv', sep = '.'),\u001b[K\n",
      "          col.names = TRUE,\u001b[K\n",
      "          row.names = FALSE,\u001b[K\n",
      "          sep = '\t',\u001b[K\n",
      "          quote = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Save model to file\u001b[K\n",
      "  \u001b[K\n",
      "  write(model, file=paste(opt$output_prefix, 'deseq2.model.txt', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## R SESSION INFO                             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  sink(paste(opt$output_prefix, \"R_sessionInfo.log\", sep = '.'))\u001b[K\n",
      "  print(sessionInfo())\u001b[K\n",
      "  sink()\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## VERSIONS FILE                              ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  r.version <- strsplit(version[['version.string']], ' ')[[1]][3]\u001b[K\n",
      "  deseq2.version <- as.character(packageVersion('DESeq2'))\u001b[K\n",
      "  \u001b[K\n",
      "  writeLines(\u001b[K\n",
      "      c(\u001b[K\n",
      "          '\"NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL\":',\u001b[K\n",
      "          paste('    r-base:', r.version),\u001b[K\n",
      "          paste('    bioconductor-deseq2:', deseq2.version)\u001b[K\n",
      "      ),\u001b[K\n",
      "  'versions.yml')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "\u001b[K\n",
      "Command exit status:\u001b[K\n",
      "  125\u001b[K\n",
      "\u001b[K\n",
      "Command output:\u001b[K\n",
      "  (empty)\u001b[K\n",
      "\u001b[K\n",
      "Command error:\u001b[K\n",
      "  Loading required package: matrixStats\u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘MatrixGenerics’\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\u001b[K\n",
      "      colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\u001b[K\n",
      "      colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\u001b[K\n",
      "      colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\u001b[K\n",
      "      colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\u001b[K\n",
      "      colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\u001b[K\n",
      "      colWeightedMeans, colWeightedMedians, colWeightedSds,\u001b[K\n",
      "      colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\u001b[K\n",
      "      rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\u001b[K\n",
      "      rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\u001b[K\n",
      "      rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\u001b[K\n",
      "      rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\u001b[K\n",
      "      rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\u001b[K\n",
      "      rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\u001b[K\n",
      "      rowWeightedSds, rowWeightedVars\u001b[K\n",
      "  \u001b[K\n",
      "  Loading required package: Biobase\u001b[K\n",
      "  Welcome to Bioconductor\u001b[K\n",
      "  \u001b[K\n",
      "      Vignettes contain introductory material; view with\u001b[K\n",
      "      'browseVignettes()'. To cite Bioconductor, see\u001b[K\n",
      "      'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\u001b[K\n",
      "  \u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘Biobase’\u001b[K\n",
      "  \u001b[K\n",
      "  The following object is masked from ‘package:MatrixGenerics’:\u001b[K\n",
      "  \u001b[K\n",
      "      rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      anyMissing, rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  converting counts to integer mode\u001b[K\n",
      "  estimating size factors\u001b[K\n",
      "  estimating dispersions\u001b[K\n",
      "  gene-wise dispersion estimates: 1 workers\u001b[K\n",
      "  mean-dispersion relationship\u001b[K\n",
      "  final dispersion estimates, fitting model and testing: 1 workers\u001b[K\n",
      "  using 'ashr' for LFC shrinkage. If used in published research, please cite:\u001b[K\n",
      "      Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\u001b[K\n",
      "      https://doi.org/10.1093/biostatistics/kxw041\u001b[K\n",
      "  time=\"2025-10-01T11:42:37+02:00\" level=error msg=\"error waiting for container: unexpected EOF\"\u001b[K\n",
      "\u001b[K\n",
      "Work dir:\u001b[K\n",
      "  /home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_03/work/a4/194845d6f6d9a0fc47f929230de056\u001b[K\n",
      "\u001b[K\n",
      "Container:\u001b[K\n",
      "  quay.io/biocontainers/bioconductor-deseq2:1.34.0--r41hc247a5b_3\u001b[K\n",
      "\u001b[K\n",
      "Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`\u001b[K\n",
      "\u001b[K\n",
      " -- Check '.nextflow.log' file for details\u001b[39m\u001b[K\n",
      "\u001b[633A\n",
      "\u001b[2mexecutor >  local (4)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m8c/aaa19e\u001b[0;2m] \u001b[0;2m\u001b[mNFC…_MATRIXFILTER\u001b[33;2m (\u001b[0;33m[id:study]\u001b[2m)\u001b[m\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM\u001b[2m |\u001b[m 0 of 1\u001b[K\n",
      "\u001b[2m[\u001b[0;34m33/18b350\u001b[0;2m] \u001b[0;2mNFC…rget:Sham_Sal, blocking:\u001b[m])\u001b[2m |\u001b[m 0 of 3\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "Execution cancelled -- Finishing pending tasks before exit\u001b[K\n",
      "\u001b[31mERROR ~ Error executing process > 'NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])'\u001b[K\n",
      "\u001b[K\n",
      "Caused by:\u001b[K\n",
      "  Process `NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])` terminated with an error exit status (125)\u001b[K\n",
      "\u001b[K\n",
      "\u001b[K\n",
      "Command executed [/home/marcelullose/.nextflow/assets/nf-core/differentialabundance/modules/nf-core/deseq2/differential/templates/deseq_de.R]:\u001b[K\n",
      "\u001b[K\n",
      "  #!/usr/bin/env Rscript\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Functions                                  ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  #' Check for Non-Empty, Non-Whitespace String\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' This function checks if the input is non-NULL and contains more than just whitespace.\u001b[K\n",
      "  #' It returns TRUE if the input is a non-empty, non-whitespace string, and FALSE otherwise.\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param input A variable to check.\u001b[K\n",
      "  #' @return A logical value: TRUE if the input is a valid, non-empty, non-whitespace string; FALSE otherwise.\u001b[K\n",
      "  #' @examples\u001b[K\n",
      "  #' is_valid_string(\"Hello World\") # Returns TRUE\u001b[K\n",
      "  #' is_valid_string(\"   \")         # Returns FALSE\u001b[K\n",
      "  #' is_valid_string(NULL)          # Returns FALSE\u001b[K\n",
      "  \u001b[K\n",
      "  is_valid_string <- function(input) {\u001b[K\n",
      "      !is.null(input) && nzchar(trimws(input))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Parse out options from a string without recourse to optparse\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param x Long-form argument list like --opt1 val1 --opt2 val2\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return named list of options and values similar to optparse\u001b[K\n",
      "  \u001b[K\n",
      "  parse_args <- function(x){\u001b[K\n",
      "      args_list <- unlist(strsplit(x, ' ?--')[[1]])[-1]\u001b[K\n",
      "      args_vals <- lapply(args_list, function(x) scan(text=x, what='character', quiet = TRUE))\u001b[K\n",
      "  \u001b[K\n",
      "      # Ensure the option vectors are length 2 (key/ value) to catch empty ones\u001b[K\n",
      "      args_vals <- lapply(args_vals, function(z){ length(z) <- 2; z})\u001b[K\n",
      "  \u001b[K\n",
      "      parsed_args <- structure(lapply(args_vals, function(x) x[2]), names = lapply(args_vals, function(x) x[1]))\u001b[K\n",
      "      parsed_args[! is.na(parsed_args)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Flexibly read CSV or TSV files\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param file Input file\u001b[K\n",
      "  #' @param header Passed to read.delim()\u001b[K\n",
      "  #' @param row.names Passed to read.delim()\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  read_delim_flexible <- function(file, header = TRUE, row.names = NULL, check.names = TRUE){\u001b[K\n",
      "  \u001b[K\n",
      "      ext <- tolower(tail(strsplit(basename(file), split = \"\\\\.\")[[1]], 1))\u001b[K\n",
      "  \u001b[K\n",
      "      if (ext == \"tsv\" || ext == \"txt\") {\u001b[K\n",
      "          separator <- \"\\t\"\u001b[K\n",
      "      } else if (ext == \"csv\") {\u001b[K\n",
      "          separator <- \",\"\u001b[K\n",
      "      } else {\u001b[K\n",
      "          stop(paste(\"Unknown separator for\", ext))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      read.delim(\u001b[K\n",
      "          file,\u001b[K\n",
      "          sep = separator,\u001b[K\n",
      "          header = header,\u001b[K\n",
      "          row.names = row.names,\u001b[K\n",
      "          check.names = check.names\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Round numeric dataframe columns to fixed decimal places by applying\u001b[K\n",
      "  #' formatting and converting back to numerics\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param dataframe A data frame\u001b[K\n",
      "  #' @param columns Which columns to round (assumes all of them by default)\u001b[K\n",
      "  #' @param digits How many decimal places to round to?\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  round_dataframe_columns <- function(df, columns = NULL, digits = 8){\u001b[K\n",
      "      if (is.null(columns)){\u001b[K\n",
      "          columns <- colnames(df)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      df[,columns] <- format(\u001b[K\n",
      "          data.frame(df[, columns], check.names = FALSE),\u001b[K\n",
      "          nsmall = digits\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "      # Convert columns back to numeric\u001b[K\n",
      "  \u001b[K\n",
      "      for (c in columns) {\u001b[K\n",
      "          df[[c]][grep(\"^ *NA$\", df[[c]])] <- NA\u001b[K\n",
      "          df[[c]] <- as.numeric(df[[c]])\u001b[K\n",
      "      }\u001b[K\n",
      "      df\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## PARSE PARAMETERS FROM NEXTFLOW             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  # I've defined these in a single array like this so that we could go back to an\u001b[K\n",
      "  # optparse-driven method in future with module bin/ directories, rather than\u001b[K\n",
      "  # the template\u001b[K\n",
      "  \u001b[K\n",
      "  # Set defaults and classes\u001b[K\n",
      "  \u001b[K\n",
      "  opt <- list(\u001b[K\n",
      "      output_prefix = ifelse('Sham_oxy_vs_Sham_Sal' == 'null', 'Sham_oxy_vs_Sham_Sal', 'Sham_oxy_vs_Sham_Sal'),\u001b[K\n",
      "      count_file = 'study.filtered.tsv',\u001b[K\n",
      "      sample_file = 'samplesheet_diffabundance.sample_metadata.tsv',\u001b[K\n",
      "      contrast_variable = 'Condition',\u001b[K\n",
      "      reference_level = 'Sham_oxy',\u001b[K\n",
      "      target_level = 'Sham_Sal',\u001b[K\n",
      "      blocking_variables = NULL,\u001b[K\n",
      "      control_genes_file = '',\u001b[K\n",
      "      transcript_lengths_file = '',\u001b[K\n",
      "      sizefactors_from_controls = FALSE,\u001b[K\n",
      "      gene_id_col = \"gene_id\",\u001b[K\n",
      "      sample_id_col = \"experiment_accession\",\u001b[K\n",
      "      subset_to_contrast_samples = FALSE,\u001b[K\n",
      "      exclude_samples_col = NULL,\u001b[K\n",
      "      exclude_samples_values = NULL,\u001b[K\n",
      "      test = \"Wald\",\u001b[K\n",
      "      fit_type = \"parametric\",\u001b[K\n",
      "      sf_type = 'ratio',\u001b[K\n",
      "      min_replicates_for_replace = 7,\u001b[K\n",
      "      use_t = FALSE,\u001b[K\n",
      "      lfc_threshold = 0,\u001b[K\n",
      "      alt_hypothesis = 'greaterAbs',\u001b[K\n",
      "      independent_filtering = TRUE,\u001b[K\n",
      "      p_adjust_method = 'BH',\u001b[K\n",
      "      alpha = 0.1,\u001b[K\n",
      "      minmu = 0.5,\u001b[K\n",
      "      vs_method = 'vst', # 'rlog', 'vst', or 'rlog,vst'\u001b[K\n",
      "      shrink_lfc = TRUE,\u001b[K\n",
      "      cores = 1,\u001b[K\n",
      "      vs_blind = TRUE,\u001b[K\n",
      "      vst_nsub = 1000\u001b[K\n",
      "  )\u001b[K\n",
      "  opt_types <- lapply(opt, class)\u001b[K\n",
      "  \u001b[K\n",
      "  # Apply parameter overrides\u001b[K\n",
      "  \u001b[K\n",
      "  args_opt <- parse_args('--gene_id_col \"gene_id\" --sample_id_col \"sample\" --test Wald --fit_type parametric --sizefactors_from_controls false --sf_type ratio --min_replicates_for_replace 7 --use_t false --lfc_threshold 0 --alt_hypothesis greaterAbs --independent_filtering true --p_adjust_method BH --alpha 0.1 --minmu 0.5 --vs_method vst --vs_blind true --vst_nsub 1000 --shrink_lfc true --cores 1 --subset_to_contrast_samples false --blocking_variables')\u001b[K\n",
      "  for ( ao in names(args_opt)){\u001b[K\n",
      "      if (! ao %in% names(opt)){\u001b[K\n",
      "          stop(paste(\"Invalid option:\", ao))\u001b[K\n",
      "      }else{\u001b[K\n",
      "  \u001b[K\n",
      "          # Preserve classes from defaults where possible\u001b[K\n",
      "          if (! is.null(opt[[ao]])){\u001b[K\n",
      "              args_opt[[ao]] <- as(args_opt[[ao]], opt_types[[ao]])\u001b[K\n",
      "          }\u001b[K\n",
      "          opt[[ao]] <- args_opt[[ao]]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check if required parameters have been provided\u001b[K\n",
      "  \u001b[K\n",
      "  required_opts <- c('contrast_variable', 'reference_level', 'target_level', 'output_prefix')\u001b[K\n",
      "  missing <- required_opts[!unlist(lapply(opt[required_opts], is_valid_string)) | !required_opts %in% names(opt)]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing) > 0){\u001b[K\n",
      "      stop(paste(\"Missing required options:\", paste(missing, collapse=', ')))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check file inputs are valid\u001b[K\n",
      "  \u001b[K\n",
      "  for (file_input in c('count_file', 'sample_file')){\u001b[K\n",
      "      if (! is_valid_string(opt[[file_input]])) {\u001b[K\n",
      "          stop(paste(\"Please provide\", file_input), call. = FALSE)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      if (! file.exists(opt[[file_input]])){\u001b[K\n",
      "          stop(paste0('Value of ', file_input, ': ', opt[[file_input]], ' is not a valid file'))\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Finish loading libraries                   ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  library(DESeq2)\u001b[K\n",
      "  library(BiocParallel)\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## READ IN COUNTS FILE AND SAMPLE METADATA    ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  count.table <-\u001b[K\n",
      "      read_delim_flexible(\u001b[K\n",
      "          file = opt$count_file,\u001b[K\n",
      "          header = TRUE,\u001b[K\n",
      "          row.names = opt$gene_id_col,\u001b[K\n",
      "          check.names = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  sample.sheet <- read_delim_flexible(file = opt$sample_file)\u001b[K\n",
      "  \u001b[K\n",
      "  # Deal with spaces that may be in sample column\u001b[K\n",
      "  opt$sample_id_col <- make.names(opt$sample_id_col)\u001b[K\n",
      "  \u001b[K\n",
      "  if (! opt$sample_id_col %in% colnames(sample.sheet)){\u001b[K\n",
      "      stop(paste0(\"Specified sample ID column '\", opt$sample_id_col, \"' is not in the sample sheet\"))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Sample sheet can have duplicate rows for multiple sequencing runs, so uniqify\u001b[K\n",
      "  # before assigning row names\u001b[K\n",
      "  \u001b[K\n",
      "  sample.sheet <- sample.sheet[! duplicated(sample.sheet[[opt$sample_id_col]]), ]\u001b[K\n",
      "  rownames(sample.sheet) <- sample.sheet[[opt$sample_id_col]]\u001b[K\n",
      "  \u001b[K\n",
      "  # Check that all samples specified in the input sheet are present in the counts\u001b[K\n",
      "  # table. Assuming they are, subset and sort the count table to match the sample\u001b[K\n",
      "  # sheet\u001b[K\n",
      "  \u001b[K\n",
      "  missing_samples <-\u001b[K\n",
      "      sample.sheet[!rownames(sample.sheet) %in% colnames(count.table), opt$sample_id_col]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing_samples) > 0) {\u001b[K\n",
      "      stop(paste(\u001b[K\n",
      "          length(missing_samples),\u001b[K\n",
      "          'specified samples missing from count table:',\u001b[K\n",
      "          paste(missing_samples, collapse = ',')\u001b[K\n",
      "      ))\u001b[K\n",
      "  } else{\u001b[K\n",
      "      # Save any non-count data, will gene metadata etc we might need later\u001b[K\n",
      "      noncount.table <-\u001b[K\n",
      "          count.table[, !colnames(count.table) %in% rownames(sample.sheet), drop = FALSE]\u001b[K\n",
      "      count.table <- count.table[, rownames(sample.sheet)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## CHECK CONTRAST SPECIFICATION               ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast_variable <- make.names(opt$contrast_variable)\u001b[K\n",
      "  blocking.vars <- c()\u001b[K\n",
      "  \u001b[K\n",
      "  if (!contrast_variable %in% colnames(sample.sheet)) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste0(\u001b[K\n",
      "          'Chosen contrast variable \\\"',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          '\\\" not in sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (any(!c(opt$reflevel, opt$treatlevel) %in% sample.sheet[[contrast_variable]])) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste(\u001b[K\n",
      "          'Please choose reference and treatment levels that are present in the',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          'column of the sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      blocking.vars = make.names(unlist(strsplit(opt$blocking_variables, split = ';')))\u001b[K\n",
      "      if (!all(blocking.vars %in% colnames(sample.sheet))) {\u001b[K\n",
      "          missing_block <- paste(blocking.vars[! blocking.vars %in% colnames(sample.sheet)], collapse = ',')\u001b[K\n",
      "          stop(\u001b[K\n",
      "              paste(\u001b[K\n",
      "                  'Blocking variables', missing_block,\u001b[K\n",
      "                  'do not correspond to sample sheet columns.'\u001b[K\n",
      "              )\u001b[K\n",
      "          )\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, subset to only the samples involved in the contrast\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$subset_to_contrast_samples){\u001b[K\n",
      "      sample_selector <- sample.sheet[[contrast_variable]] %in% c(opt$target_level, opt$reference_level)\u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, remove samples with specified values in a given field (probably\u001b[K\n",
      "  # don't use this as well as the above)\u001b[K\n",
      "  \u001b[K\n",
      "  if ((is_valid_string(opt$exclude_samples_col)) && (is_valid_string(opt$exclude_samples_values))){\u001b[K\n",
      "      exclude_values = unlist(strsplit(opt$exclude_samples_values, split = ';'))\u001b[K\n",
      "  \u001b[K\n",
      "      if (! opt$exclude_samples_col %in% colnames(sample.sheet)){\u001b[K\n",
      "          stop(paste(opt$exclude_samples_col, ' specified to subset samples is not a valid sample sheet column'))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      print(paste0('Excluding samples with values of ', opt$exclude_samples_values, ' in ', opt$exclude_samples_col))\u001b[K\n",
      "      sample_selector <- ! sample.sheet[[opt$exclude_samples_col]] %in% exclude_values\u001b[K\n",
      "  \u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Now specify the model. Use cell-means style so we can be explicit with the\u001b[K\n",
      "  # contrasts\u001b[K\n",
      "  \u001b[K\n",
      "  model <- '~ 0'\u001b[K\n",
      "  \u001b[K\n",
      "  if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      model <- paste(model, paste(blocking.vars, collapse = ' + '), sep=' + ')\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Make sure all the appropriate variables are factors\u001b[K\n",
      "  \u001b[K\n",
      "  for (v in c(blocking.vars, contrast_variable)) {\u001b[K\n",
      "      sample.sheet[[v]] <- as.factor(sample.sheet[[v]])\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Variable of interest goes last, see\u001b[K\n",
      "  # https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs\u001b[K\n",
      "  \u001b[K\n",
      "  model <- paste(model, contrast_variable, sep = ' + ')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Run DESeq2 processes                       ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != ''){\u001b[K\n",
      "      control_genes <- readLines(opt$control_genes_file)\u001b[K\n",
      "      if (! opt$sizefactors_from_controls){\u001b[K\n",
      "          count.table <- count.table[setdiff(rownames(count.table), control_genes),]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeqDataSetFromMatrix(\u001b[K\n",
      "      countData = round(count.table),\u001b[K\n",
      "      colData = sample.sheet,\u001b[K\n",
      "      design = as.formula(model)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Build in transcript lengths. Copying what tximport does here:\u001b[K\n",
      "  # https://github.com/thelovelab/DESeq2/blob/6947d5bc629015fb8ffb2453a91b71665a164483/R/AllClasses.R#L409\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      lengths <-\u001b[K\n",
      "          read_delim_flexible(\u001b[K\n",
      "              file = opt$transcript_lengths_file,\u001b[K\n",
      "              header = TRUE,\u001b[K\n",
      "              row.names = opt$gene_id_col,\u001b[K\n",
      "              check.names = FALSE\u001b[K\n",
      "          )\u001b[K\n",
      "      lengths <- lengths[rownames(count.table), colnames(count.table)]\u001b[K\n",
      "      dimnames(lengths) <- dimnames(dds)\u001b[K\n",
      "      assays(dds)[[\"avgTxLength\"]] <- lengths\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != '' && opt$sizefactors_from_controls){\u001b[K\n",
      "      print(paste('Estimating size factors using', length(control_genes), 'control genes'))\u001b[K\n",
      "      dds <- estimateSizeFactors(dds, controlGenes=rownames(count.table) %in% control_genes)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeq(\u001b[K\n",
      "      dds,\u001b[K\n",
      "      test = opt$test,\u001b[K\n",
      "      fitType = opt$fit_type,\u001b[K\n",
      "      minReplicatesForReplace = opt$min_replicates_for_replace,\u001b[K\n",
      "      useT = opt$use_t,\u001b[K\n",
      "      sfType = opt$sf_type,\u001b[K\n",
      "      parallel=TRUE, BPPARAM=MulticoreParam(opt$cores)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  comp.results <-\u001b[K\n",
      "      results(\u001b[K\n",
      "          dds,\u001b[K\n",
      "          lfcThreshold = opt$lfc_threshold,\u001b[K\n",
      "          altHypothesis = opt$alt_hypothesis,\u001b[K\n",
      "          independentFiltering = opt$independent_filtering,\u001b[K\n",
      "          alpha = opt$alpha,\u001b[K\n",
      "          pAdjustMethod = opt$p_adjust_method,\u001b[K\n",
      "          minmu = opt$minmu,\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$shrink_lfc){\u001b[K\n",
      "      comp.results <- lfcShrink(dds,\u001b[K\n",
      "          type = 'ashr',\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # See https://support.bioconductor.org/p/97676/\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      size_factors = estimateSizeFactorsForMatrix(counts(dds) / assays(dds)[[\"avgTxLength\"]])\u001b[K\n",
      "  }else {\u001b[K\n",
      "      size_factors = sizeFactors(dds)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Generate outputs                           ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast.name <-\u001b[K\n",
      "      paste(opt$target_level, opt$reference_level, sep = \"_vs_\")\u001b[K\n",
      "  cat(\"Saving results for \", contrast.name, \" ...\u001b[K\n",
      "  \", sep = \"\")\u001b[K\n",
      "  \u001b[K\n",
      "  # Differential expression table- note very limited rounding for consistency of\u001b[K\n",
      "  # results\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(comp.results)), opt$gene_id_col),\u001b[K\n",
      "      round_dataframe_columns(\u001b[K\n",
      "          data.frame(comp.results[, !(colnames(comp.results) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "      )\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.results.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Dispersion plot\u001b[K\n",
      "  \u001b[K\n",
      "  png(\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.dispersion.png', sep = '.'),\u001b[K\n",
      "      width = 600,\u001b[K\n",
      "      height = 600\u001b[K\n",
      "  )\u001b[K\n",
      "  plotDispEsts(dds)\u001b[K\n",
      "  dev.off()\u001b[K\n",
      "  \u001b[K\n",
      "  # R object for other processes to use\u001b[K\n",
      "  \u001b[K\n",
      "  saveRDS(dds, file = paste(opt$output_prefix, 'dds.rld.rds', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  # Size factors\u001b[K\n",
      "  \u001b[K\n",
      "  sf_df = data.frame(\u001b[K\n",
      "      sample = names(size_factors),\u001b[K\n",
      "      data.frame(size_factors, check.names = FALSE),\u001b[K\n",
      "      check.names = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  colnames(sf_df) <- c('sample', 'sizeFactor')\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      sf_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.sizefactors.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Write specified matrices\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "      data.frame(counts(dds, normalized = TRUE)[, !(colnames(counts(dds, normalized = TRUE)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'normalised_counts.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Note very limited rounding for consistency of results\u001b[K\n",
      "  \u001b[K\n",
      "  for (vs_method_name in strsplit(opt$vs_method, ',')){\u001b[K\n",
      "      if (vs_method_name == 'vst'){\u001b[K\n",
      "          vs_mat <- vst(dds, blind = opt$vs_blind, nsub = opt$vst_nsub)\u001b[K\n",
      "      }else if (vs_method_name == 'rlog'){\u001b[K\n",
      "          vs_mat <- rlog(dds, blind = opt$vs_blind, fitType = opt$fit_type)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      # Again apply the slight rounding and then restore numeric\u001b[K\n",
      "  \u001b[K\n",
      "      out_df <- cbind(\u001b[K\n",
      "          setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "          round_dataframe_columns(\u001b[K\n",
      "              data.frame(assay(vs_mat)[, !(colnames(assay(vs_mat)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "      write.table(\u001b[K\n",
      "          out_df,\u001b[K\n",
      "          file = paste(opt$output_prefix, vs_method_name, 'tsv', sep = '.'),\u001b[K\n",
      "          col.names = TRUE,\u001b[K\n",
      "          row.names = FALSE,\u001b[K\n",
      "          sep = '\t',\u001b[K\n",
      "          quote = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Save model to file\u001b[K\n",
      "  \u001b[K\n",
      "  write(model, file=paste(opt$output_prefix, 'deseq2.model.txt', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## R SESSION INFO                             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  sink(paste(opt$output_prefix, \"R_sessionInfo.log\", sep = '.'))\u001b[K\n",
      "  print(sessionInfo())\u001b[K\n",
      "  sink()\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## VERSIONS FILE                              ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  r.version <- strsplit(version[['version.string']], ' ')[[1]][3]\u001b[K\n",
      "  deseq2.version <- as.character(packageVersion('DESeq2'))\u001b[K\n",
      "  \u001b[K\n",
      "  writeLines(\u001b[K\n",
      "      c(\u001b[K\n",
      "          '\"NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL\":',\u001b[K\n",
      "          paste('    r-base:', r.version),\u001b[K\n",
      "          paste('    bioconductor-deseq2:', deseq2.version)\u001b[K\n",
      "      ),\u001b[K\n",
      "  'versions.yml')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "\u001b[K\n",
      "Command exit status:\u001b[K\n",
      "  125\u001b[K\n",
      "\u001b[K\n",
      "Command output:\u001b[K\n",
      "  (empty)\u001b[K\n",
      "\u001b[K\n",
      "Command error:\u001b[K\n",
      "  Loading required package: matrixStats\u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘MatrixGenerics’\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\u001b[K\n",
      "      colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\u001b[K\n",
      "      colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\u001b[K\n",
      "      colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\u001b[K\n",
      "      colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\u001b[K\n",
      "      colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\u001b[K\n",
      "      colWeightedMeans, colWeightedMedians, colWeightedSds,\u001b[K\n",
      "      colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\u001b[K\n",
      "      rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\u001b[K\n",
      "      rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\u001b[K\n",
      "      rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\u001b[K\n",
      "      rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\u001b[K\n",
      "      rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\u001b[K\n",
      "      rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\u001b[K\n",
      "      rowWeightedSds, rowWeightedVars\u001b[K\n",
      "  \u001b[K\n",
      "  Loading required package: Biobase\u001b[K\n",
      "  Welcome to Bioconductor\u001b[K\n",
      "  \u001b[K\n",
      "      Vignettes contain introductory material; view with\u001b[K\n",
      "      'browseVignettes()'. To cite Bioconductor, see\u001b[K\n",
      "      'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\u001b[K\n",
      "  \u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘Biobase’\u001b[K\n",
      "  \u001b[K\n",
      "  The following object is masked from ‘package:MatrixGenerics’:\u001b[K\n",
      "  \u001b[K\n",
      "      rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      anyMissing, rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  converting counts to integer mode\u001b[K\n",
      "  estimating size factors\u001b[K\n",
      "  estimating dispersions\u001b[K\n",
      "  gene-wise dispersion estimates: 1 workers\u001b[K\n",
      "  mean-dispersion relationship\u001b[K\n",
      "  final dispersion estimates, fitting model and testing: 1 workers\u001b[K\n",
      "  using 'ashr' for LFC shrinkage. If used in published research, please cite:\u001b[K\n",
      "      Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\u001b[K\n",
      "      https://doi.org/10.1093/biostatistics/kxw041\u001b[K\n",
      "  time=\"2025-10-01T11:42:37+02:00\" level=error msg=\"error waiting for container: unexpected EOF\"\u001b[K\n",
      "\u001b[K\n",
      "Work dir:\u001b[K\n",
      "  /home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_03/work/a4/194845d6f6d9a0fc47f929230de056\u001b[K\n",
      "\u001b[K\n",
      "Container:\u001b[K\n",
      "  quay.io/biocontainers/bioconductor-deseq2:1.34.0--r41hc247a5b_3\u001b[K\n",
      "\u001b[K\n",
      "Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`\u001b[K\n",
      "\u001b[K\n",
      " -- Check '.nextflow.log' file for details\u001b[39m\u001b[K\n",
      "\n",
      "\u001b[633A\n",
      "\u001b[2mexecutor >  local (4)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m31/f446d0\u001b[0;2m] \u001b[0;2m\u001b[mNFC…lesheet_diffabundance.csv)\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m8c/aaa19e\u001b[0;2m] \u001b[0;2m\u001b[mNFC…_MATRIXFILTER\u001b[33;2m (\u001b[0;33m[id:study]\u001b[2m)\u001b[m\u001b[2m |\u001b[m 1 of 1\u001b[32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NTIALABUNDANCE:\u001b[mDESEQ2_NORM\u001b[2m |\u001b[m 0 of 1\u001b[K\n",
      "\u001b[2m[\u001b[0;34m33/18b350\u001b[0;2m] \u001b[0;2mNFC…rget:Sham_Sal, blocking:\u001b[m])\u001b[2m |\u001b[m 0 of 3\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mFILTER_DIFFTABLE -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…ABUNDANCE:\u001b[mPLOT_EXPLORATORY -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mPLOT_DIFFERENTIAL -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TIALABUNDANCE:\u001b[mSHINYNGS_APP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…BUNDANCE:\u001b[mRMARKDOWNNOTEBOOK -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…UNDANCE:\u001b[mMAKE_REPORT_BUNDLE -\u001b[K\n",
      "Execution cancelled -- Finishing pending tasks before exit\u001b[K\n",
      "\u001b[31mERROR ~ Error executing process > 'NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])'\u001b[K\n",
      "\u001b[K\n",
      "Caused by:\u001b[K\n",
      "  Process `NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL ([id:Sham_oxy_vs_Sham_Sal, variable:Condition, reference:Sham_oxy, target:Sham_Sal, blocking:])` terminated with an error exit status (125)\u001b[K\n",
      "\u001b[K\n",
      "\u001b[K\n",
      "Command executed [/home/marcelullose/.nextflow/assets/nf-core/differentialabundance/modules/nf-core/deseq2/differential/templates/deseq_de.R]:\u001b[K\n",
      "\u001b[K\n",
      "  #!/usr/bin/env Rscript\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Functions                                  ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  #' Check for Non-Empty, Non-Whitespace String\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' This function checks if the input is non-NULL and contains more than just whitespace.\u001b[K\n",
      "  #' It returns TRUE if the input is a non-empty, non-whitespace string, and FALSE otherwise.\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param input A variable to check.\u001b[K\n",
      "  #' @return A logical value: TRUE if the input is a valid, non-empty, non-whitespace string; FALSE otherwise.\u001b[K\n",
      "  #' @examples\u001b[K\n",
      "  #' is_valid_string(\"Hello World\") # Returns TRUE\u001b[K\n",
      "  #' is_valid_string(\"   \")         # Returns FALSE\u001b[K\n",
      "  #' is_valid_string(NULL)          # Returns FALSE\u001b[K\n",
      "  \u001b[K\n",
      "  is_valid_string <- function(input) {\u001b[K\n",
      "      !is.null(input) && nzchar(trimws(input))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Parse out options from a string without recourse to optparse\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param x Long-form argument list like --opt1 val1 --opt2 val2\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return named list of options and values similar to optparse\u001b[K\n",
      "  \u001b[K\n",
      "  parse_args <- function(x){\u001b[K\n",
      "      args_list <- unlist(strsplit(x, ' ?--')[[1]])[-1]\u001b[K\n",
      "      args_vals <- lapply(args_list, function(x) scan(text=x, what='character', quiet = TRUE))\u001b[K\n",
      "  \u001b[K\n",
      "      # Ensure the option vectors are length 2 (key/ value) to catch empty ones\u001b[K\n",
      "      args_vals <- lapply(args_vals, function(z){ length(z) <- 2; z})\u001b[K\n",
      "  \u001b[K\n",
      "      parsed_args <- structure(lapply(args_vals, function(x) x[2]), names = lapply(args_vals, function(x) x[1]))\u001b[K\n",
      "      parsed_args[! is.na(parsed_args)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Flexibly read CSV or TSV files\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param file Input file\u001b[K\n",
      "  #' @param header Passed to read.delim()\u001b[K\n",
      "  #' @param row.names Passed to read.delim()\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  read_delim_flexible <- function(file, header = TRUE, row.names = NULL, check.names = TRUE){\u001b[K\n",
      "  \u001b[K\n",
      "      ext <- tolower(tail(strsplit(basename(file), split = \"\\\\.\")[[1]], 1))\u001b[K\n",
      "  \u001b[K\n",
      "      if (ext == \"tsv\" || ext == \"txt\") {\u001b[K\n",
      "          separator <- \"\\t\"\u001b[K\n",
      "      } else if (ext == \"csv\") {\u001b[K\n",
      "          separator <- \",\"\u001b[K\n",
      "      } else {\u001b[K\n",
      "          stop(paste(\"Unknown separator for\", ext))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      read.delim(\u001b[K\n",
      "          file,\u001b[K\n",
      "          sep = separator,\u001b[K\n",
      "          header = header,\u001b[K\n",
      "          row.names = row.names,\u001b[K\n",
      "          check.names = check.names\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  #' Round numeric dataframe columns to fixed decimal places by applying\u001b[K\n",
      "  #' formatting and converting back to numerics\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @param dataframe A data frame\u001b[K\n",
      "  #' @param columns Which columns to round (assumes all of them by default)\u001b[K\n",
      "  #' @param digits How many decimal places to round to?\u001b[K\n",
      "  #'\u001b[K\n",
      "  #' @return output Data frame\u001b[K\n",
      "  \u001b[K\n",
      "  round_dataframe_columns <- function(df, columns = NULL, digits = 8){\u001b[K\n",
      "      if (is.null(columns)){\u001b[K\n",
      "          columns <- colnames(df)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      df[,columns] <- format(\u001b[K\n",
      "          data.frame(df[, columns], check.names = FALSE),\u001b[K\n",
      "          nsmall = digits\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "      # Convert columns back to numeric\u001b[K\n",
      "  \u001b[K\n",
      "      for (c in columns) {\u001b[K\n",
      "          df[[c]][grep(\"^ *NA$\", df[[c]])] <- NA\u001b[K\n",
      "          df[[c]] <- as.numeric(df[[c]])\u001b[K\n",
      "      }\u001b[K\n",
      "      df\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## PARSE PARAMETERS FROM NEXTFLOW             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  # I've defined these in a single array like this so that we could go back to an\u001b[K\n",
      "  # optparse-driven method in future with module bin/ directories, rather than\u001b[K\n",
      "  # the template\u001b[K\n",
      "  \u001b[K\n",
      "  # Set defaults and classes\u001b[K\n",
      "  \u001b[K\n",
      "  opt <- list(\u001b[K\n",
      "      output_prefix = ifelse('Sham_oxy_vs_Sham_Sal' == 'null', 'Sham_oxy_vs_Sham_Sal', 'Sham_oxy_vs_Sham_Sal'),\u001b[K\n",
      "      count_file = 'study.filtered.tsv',\u001b[K\n",
      "      sample_file = 'samplesheet_diffabundance.sample_metadata.tsv',\u001b[K\n",
      "      contrast_variable = 'Condition',\u001b[K\n",
      "      reference_level = 'Sham_oxy',\u001b[K\n",
      "      target_level = 'Sham_Sal',\u001b[K\n",
      "      blocking_variables = NULL,\u001b[K\n",
      "      control_genes_file = '',\u001b[K\n",
      "      transcript_lengths_file = '',\u001b[K\n",
      "      sizefactors_from_controls = FALSE,\u001b[K\n",
      "      gene_id_col = \"gene_id\",\u001b[K\n",
      "      sample_id_col = \"experiment_accession\",\u001b[K\n",
      "      subset_to_contrast_samples = FALSE,\u001b[K\n",
      "      exclude_samples_col = NULL,\u001b[K\n",
      "      exclude_samples_values = NULL,\u001b[K\n",
      "      test = \"Wald\",\u001b[K\n",
      "      fit_type = \"parametric\",\u001b[K\n",
      "      sf_type = 'ratio',\u001b[K\n",
      "      min_replicates_for_replace = 7,\u001b[K\n",
      "      use_t = FALSE,\u001b[K\n",
      "      lfc_threshold = 0,\u001b[K\n",
      "      alt_hypothesis = 'greaterAbs',\u001b[K\n",
      "      independent_filtering = TRUE,\u001b[K\n",
      "      p_adjust_method = 'BH',\u001b[K\n",
      "      alpha = 0.1,\u001b[K\n",
      "      minmu = 0.5,\u001b[K\n",
      "      vs_method = 'vst', # 'rlog', 'vst', or 'rlog,vst'\u001b[K\n",
      "      shrink_lfc = TRUE,\u001b[K\n",
      "      cores = 1,\u001b[K\n",
      "      vs_blind = TRUE,\u001b[K\n",
      "      vst_nsub = 1000\u001b[K\n",
      "  )\u001b[K\n",
      "  opt_types <- lapply(opt, class)\u001b[K\n",
      "  \u001b[K\n",
      "  # Apply parameter overrides\u001b[K\n",
      "  \u001b[K\n",
      "  args_opt <- parse_args('--gene_id_col \"gene_id\" --sample_id_col \"sample\" --test Wald --fit_type parametric --sizefactors_from_controls false --sf_type ratio --min_replicates_for_replace 7 --use_t false --lfc_threshold 0 --alt_hypothesis greaterAbs --independent_filtering true --p_adjust_method BH --alpha 0.1 --minmu 0.5 --vs_method vst --vs_blind true --vst_nsub 1000 --shrink_lfc true --cores 1 --subset_to_contrast_samples false --blocking_variables')\u001b[K\n",
      "  for ( ao in names(args_opt)){\u001b[K\n",
      "      if (! ao %in% names(opt)){\u001b[K\n",
      "          stop(paste(\"Invalid option:\", ao))\u001b[K\n",
      "      }else{\u001b[K\n",
      "  \u001b[K\n",
      "          # Preserve classes from defaults where possible\u001b[K\n",
      "          if (! is.null(opt[[ao]])){\u001b[K\n",
      "              args_opt[[ao]] <- as(args_opt[[ao]], opt_types[[ao]])\u001b[K\n",
      "          }\u001b[K\n",
      "          opt[[ao]] <- args_opt[[ao]]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check if required parameters have been provided\u001b[K\n",
      "  \u001b[K\n",
      "  required_opts <- c('contrast_variable', 'reference_level', 'target_level', 'output_prefix')\u001b[K\n",
      "  missing <- required_opts[!unlist(lapply(opt[required_opts], is_valid_string)) | !required_opts %in% names(opt)]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing) > 0){\u001b[K\n",
      "      stop(paste(\"Missing required options:\", paste(missing, collapse=', ')))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Check file inputs are valid\u001b[K\n",
      "  \u001b[K\n",
      "  for (file_input in c('count_file', 'sample_file')){\u001b[K\n",
      "      if (! is_valid_string(opt[[file_input]])) {\u001b[K\n",
      "          stop(paste(\"Please provide\", file_input), call. = FALSE)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      if (! file.exists(opt[[file_input]])){\u001b[K\n",
      "          stop(paste0('Value of ', file_input, ': ', opt[[file_input]], ' is not a valid file'))\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Finish loading libraries                   ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  library(DESeq2)\u001b[K\n",
      "  library(BiocParallel)\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## READ IN COUNTS FILE AND SAMPLE METADATA    ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  count.table <-\u001b[K\n",
      "      read_delim_flexible(\u001b[K\n",
      "          file = opt$count_file,\u001b[K\n",
      "          header = TRUE,\u001b[K\n",
      "          row.names = opt$gene_id_col,\u001b[K\n",
      "          check.names = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  sample.sheet <- read_delim_flexible(file = opt$sample_file)\u001b[K\n",
      "  \u001b[K\n",
      "  # Deal with spaces that may be in sample column\u001b[K\n",
      "  opt$sample_id_col <- make.names(opt$sample_id_col)\u001b[K\n",
      "  \u001b[K\n",
      "  if (! opt$sample_id_col %in% colnames(sample.sheet)){\u001b[K\n",
      "      stop(paste0(\"Specified sample ID column '\", opt$sample_id_col, \"' is not in the sample sheet\"))\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Sample sheet can have duplicate rows for multiple sequencing runs, so uniqify\u001b[K\n",
      "  # before assigning row names\u001b[K\n",
      "  \u001b[K\n",
      "  sample.sheet <- sample.sheet[! duplicated(sample.sheet[[opt$sample_id_col]]), ]\u001b[K\n",
      "  rownames(sample.sheet) <- sample.sheet[[opt$sample_id_col]]\u001b[K\n",
      "  \u001b[K\n",
      "  # Check that all samples specified in the input sheet are present in the counts\u001b[K\n",
      "  # table. Assuming they are, subset and sort the count table to match the sample\u001b[K\n",
      "  # sheet\u001b[K\n",
      "  \u001b[K\n",
      "  missing_samples <-\u001b[K\n",
      "      sample.sheet[!rownames(sample.sheet) %in% colnames(count.table), opt$sample_id_col]\u001b[K\n",
      "  \u001b[K\n",
      "  if (length(missing_samples) > 0) {\u001b[K\n",
      "      stop(paste(\u001b[K\n",
      "          length(missing_samples),\u001b[K\n",
      "          'specified samples missing from count table:',\u001b[K\n",
      "          paste(missing_samples, collapse = ',')\u001b[K\n",
      "      ))\u001b[K\n",
      "  } else{\u001b[K\n",
      "      # Save any non-count data, will gene metadata etc we might need later\u001b[K\n",
      "      noncount.table <-\u001b[K\n",
      "          count.table[, !colnames(count.table) %in% rownames(sample.sheet), drop = FALSE]\u001b[K\n",
      "      count.table <- count.table[, rownames(sample.sheet)]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## CHECK CONTRAST SPECIFICATION               ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast_variable <- make.names(opt$contrast_variable)\u001b[K\n",
      "  blocking.vars <- c()\u001b[K\n",
      "  \u001b[K\n",
      "  if (!contrast_variable %in% colnames(sample.sheet)) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste0(\u001b[K\n",
      "          'Chosen contrast variable \\\"',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          '\\\" not in sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (any(!c(opt$reflevel, opt$treatlevel) %in% sample.sheet[[contrast_variable]])) {\u001b[K\n",
      "      stop(\u001b[K\n",
      "          paste(\u001b[K\n",
      "          'Please choose reference and treatment levels that are present in the',\u001b[K\n",
      "          contrast_variable,\u001b[K\n",
      "          'column of the sample sheet'\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  } else if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      blocking.vars = make.names(unlist(strsplit(opt$blocking_variables, split = ';')))\u001b[K\n",
      "      if (!all(blocking.vars %in% colnames(sample.sheet))) {\u001b[K\n",
      "          missing_block <- paste(blocking.vars[! blocking.vars %in% colnames(sample.sheet)], collapse = ',')\u001b[K\n",
      "          stop(\u001b[K\n",
      "              paste(\u001b[K\n",
      "                  'Blocking variables', missing_block,\u001b[K\n",
      "                  'do not correspond to sample sheet columns.'\u001b[K\n",
      "              )\u001b[K\n",
      "          )\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, subset to only the samples involved in the contrast\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$subset_to_contrast_samples){\u001b[K\n",
      "      sample_selector <- sample.sheet[[contrast_variable]] %in% c(opt$target_level, opt$reference_level)\u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Optionally, remove samples with specified values in a given field (probably\u001b[K\n",
      "  # don't use this as well as the above)\u001b[K\n",
      "  \u001b[K\n",
      "  if ((is_valid_string(opt$exclude_samples_col)) && (is_valid_string(opt$exclude_samples_values))){\u001b[K\n",
      "      exclude_values = unlist(strsplit(opt$exclude_samples_values, split = ';'))\u001b[K\n",
      "  \u001b[K\n",
      "      if (! opt$exclude_samples_col %in% colnames(sample.sheet)){\u001b[K\n",
      "          stop(paste(opt$exclude_samples_col, ' specified to subset samples is not a valid sample sheet column'))\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      print(paste0('Excluding samples with values of ', opt$exclude_samples_values, ' in ', opt$exclude_samples_col))\u001b[K\n",
      "      sample_selector <- ! sample.sheet[[opt$exclude_samples_col]] %in% exclude_values\u001b[K\n",
      "  \u001b[K\n",
      "      selected_samples <- sample.sheet[sample_selector, opt$sample_id_col]\u001b[K\n",
      "      count.table <- count.table[, selected_samples]\u001b[K\n",
      "      sample.sheet <- sample.sheet[selected_samples, ]\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Now specify the model. Use cell-means style so we can be explicit with the\u001b[K\n",
      "  # contrasts\u001b[K\n",
      "  \u001b[K\n",
      "  model <- '~ 0'\u001b[K\n",
      "  \u001b[K\n",
      "  if (is_valid_string(opt$blocking_variables)) {\u001b[K\n",
      "      model <- paste(model, paste(blocking.vars, collapse = ' + '), sep=' + ')\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Make sure all the appropriate variables are factors\u001b[K\n",
      "  \u001b[K\n",
      "  for (v in c(blocking.vars, contrast_variable)) {\u001b[K\n",
      "      sample.sheet[[v]] <- as.factor(sample.sheet[[v]])\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Variable of interest goes last, see\u001b[K\n",
      "  # https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs\u001b[K\n",
      "  \u001b[K\n",
      "  model <- paste(model, contrast_variable, sep = ' + ')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Run DESeq2 processes                       ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != ''){\u001b[K\n",
      "      control_genes <- readLines(opt$control_genes_file)\u001b[K\n",
      "      if (! opt$sizefactors_from_controls){\u001b[K\n",
      "          count.table <- count.table[setdiff(rownames(count.table), control_genes),]\u001b[K\n",
      "      }\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeqDataSetFromMatrix(\u001b[K\n",
      "      countData = round(count.table),\u001b[K\n",
      "      colData = sample.sheet,\u001b[K\n",
      "      design = as.formula(model)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Build in transcript lengths. Copying what tximport does here:\u001b[K\n",
      "  # https://github.com/thelovelab/DESeq2/blob/6947d5bc629015fb8ffb2453a91b71665a164483/R/AllClasses.R#L409\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      lengths <-\u001b[K\n",
      "          read_delim_flexible(\u001b[K\n",
      "              file = opt$transcript_lengths_file,\u001b[K\n",
      "              header = TRUE,\u001b[K\n",
      "              row.names = opt$gene_id_col,\u001b[K\n",
      "              check.names = FALSE\u001b[K\n",
      "          )\u001b[K\n",
      "      lengths <- lengths[rownames(count.table), colnames(count.table)]\u001b[K\n",
      "      dimnames(lengths) <- dimnames(dds)\u001b[K\n",
      "      assays(dds)[[\"avgTxLength\"]] <- lengths\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$control_genes_file != '' && opt$sizefactors_from_controls){\u001b[K\n",
      "      print(paste('Estimating size factors using', length(control_genes), 'control genes'))\u001b[K\n",
      "      dds <- estimateSizeFactors(dds, controlGenes=rownames(count.table) %in% control_genes)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  dds <- DESeq(\u001b[K\n",
      "      dds,\u001b[K\n",
      "      test = opt$test,\u001b[K\n",
      "      fitType = opt$fit_type,\u001b[K\n",
      "      minReplicatesForReplace = opt$min_replicates_for_replace,\u001b[K\n",
      "      useT = opt$use_t,\u001b[K\n",
      "      sfType = opt$sf_type,\u001b[K\n",
      "      parallel=TRUE, BPPARAM=MulticoreParam(opt$cores)\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  comp.results <-\u001b[K\n",
      "      results(\u001b[K\n",
      "          dds,\u001b[K\n",
      "          lfcThreshold = opt$lfc_threshold,\u001b[K\n",
      "          altHypothesis = opt$alt_hypothesis,\u001b[K\n",
      "          independentFiltering = opt$independent_filtering,\u001b[K\n",
      "          alpha = opt$alpha,\u001b[K\n",
      "          pAdjustMethod = opt$p_adjust_method,\u001b[K\n",
      "          minmu = opt$minmu,\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$shrink_lfc){\u001b[K\n",
      "      comp.results <- lfcShrink(dds,\u001b[K\n",
      "          type = 'ashr',\u001b[K\n",
      "          contrast = c(\u001b[K\n",
      "              contrast_variable,\u001b[K\n",
      "              c(opt$target_level, opt$reference_level)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # See https://support.bioconductor.org/p/97676/\u001b[K\n",
      "  \u001b[K\n",
      "  if (opt$transcript_lengths_file != ''){\u001b[K\n",
      "      size_factors = estimateSizeFactorsForMatrix(counts(dds) / assays(dds)[[\"avgTxLength\"]])\u001b[K\n",
      "  }else {\u001b[K\n",
      "      size_factors = sizeFactors(dds)\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## Generate outputs                           ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  contrast.name <-\u001b[K\n",
      "      paste(opt$target_level, opt$reference_level, sep = \"_vs_\")\u001b[K\n",
      "  cat(\"Saving results for \", contrast.name, \" ...\u001b[K\n",
      "  \", sep = \"\")\u001b[K\n",
      "  \u001b[K\n",
      "  # Differential expression table- note very limited rounding for consistency of\u001b[K\n",
      "  # results\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(comp.results)), opt$gene_id_col),\u001b[K\n",
      "      round_dataframe_columns(\u001b[K\n",
      "          data.frame(comp.results[, !(colnames(comp.results) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "      )\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.results.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Dispersion plot\u001b[K\n",
      "  \u001b[K\n",
      "  png(\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.dispersion.png', sep = '.'),\u001b[K\n",
      "      width = 600,\u001b[K\n",
      "      height = 600\u001b[K\n",
      "  )\u001b[K\n",
      "  plotDispEsts(dds)\u001b[K\n",
      "  dev.off()\u001b[K\n",
      "  \u001b[K\n",
      "  # R object for other processes to use\u001b[K\n",
      "  \u001b[K\n",
      "  saveRDS(dds, file = paste(opt$output_prefix, 'dds.rld.rds', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  # Size factors\u001b[K\n",
      "  \u001b[K\n",
      "  sf_df = data.frame(\u001b[K\n",
      "      sample = names(size_factors),\u001b[K\n",
      "      data.frame(size_factors, check.names = FALSE),\u001b[K\n",
      "      check.names = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  colnames(sf_df) <- c('sample', 'sizeFactor')\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      sf_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'deseq2.sizefactors.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Write specified matrices\u001b[K\n",
      "  \u001b[K\n",
      "  out_df <- cbind(\u001b[K\n",
      "      setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "      data.frame(counts(dds, normalized = TRUE)[, !(colnames(counts(dds, normalized = TRUE)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "  )\u001b[K\n",
      "  write.table(\u001b[K\n",
      "      out_df,\u001b[K\n",
      "      file = paste(opt$output_prefix, 'normalised_counts.tsv', sep = '.'),\u001b[K\n",
      "      col.names = TRUE,\u001b[K\n",
      "      row.names = FALSE,\u001b[K\n",
      "      sep = '\t',\u001b[K\n",
      "      quote = FALSE\u001b[K\n",
      "  )\u001b[K\n",
      "  \u001b[K\n",
      "  # Note very limited rounding for consistency of results\u001b[K\n",
      "  \u001b[K\n",
      "  for (vs_method_name in strsplit(opt$vs_method, ',')){\u001b[K\n",
      "      if (vs_method_name == 'vst'){\u001b[K\n",
      "          vs_mat <- vst(dds, blind = opt$vs_blind, nsub = opt$vst_nsub)\u001b[K\n",
      "      }else if (vs_method_name == 'rlog'){\u001b[K\n",
      "          vs_mat <- rlog(dds, blind = opt$vs_blind, fitType = opt$fit_type)\u001b[K\n",
      "      }\u001b[K\n",
      "  \u001b[K\n",
      "      # Again apply the slight rounding and then restore numeric\u001b[K\n",
      "  \u001b[K\n",
      "      out_df <- cbind(\u001b[K\n",
      "          setNames(data.frame(rownames(counts(dds))), opt$gene_id_col),\u001b[K\n",
      "          round_dataframe_columns(\u001b[K\n",
      "              data.frame(assay(vs_mat)[, !(colnames(assay(vs_mat)) %in% opt$gene_id_col)], check.names = FALSE)\u001b[K\n",
      "          )\u001b[K\n",
      "      )\u001b[K\n",
      "      write.table(\u001b[K\n",
      "          out_df,\u001b[K\n",
      "          file = paste(opt$output_prefix, vs_method_name, 'tsv', sep = '.'),\u001b[K\n",
      "          col.names = TRUE,\u001b[K\n",
      "          row.names = FALSE,\u001b[K\n",
      "          sep = '\t',\u001b[K\n",
      "          quote = FALSE\u001b[K\n",
      "      )\u001b[K\n",
      "  }\u001b[K\n",
      "  \u001b[K\n",
      "  # Save model to file\u001b[K\n",
      "  \u001b[K\n",
      "  write(model, file=paste(opt$output_prefix, 'deseq2.model.txt', sep = '.'))\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## R SESSION INFO                             ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  sink(paste(opt$output_prefix, \"R_sessionInfo.log\", sep = '.'))\u001b[K\n",
      "  print(sessionInfo())\u001b[K\n",
      "  sink()\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ## VERSIONS FILE                              ##\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  \u001b[K\n",
      "  r.version <- strsplit(version[['version.string']], ' ')[[1]][3]\u001b[K\n",
      "  deseq2.version <- as.character(packageVersion('DESeq2'))\u001b[K\n",
      "  \u001b[K\n",
      "  writeLines(\u001b[K\n",
      "      c(\u001b[K\n",
      "          '\"NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:DESEQ2_DIFFERENTIAL\":',\u001b[K\n",
      "          paste('    r-base:', r.version),\u001b[K\n",
      "          paste('    bioconductor-deseq2:', deseq2.version)\u001b[K\n",
      "      ),\u001b[K\n",
      "  'versions.yml')\u001b[K\n",
      "  \u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "  ################################################\u001b[K\n",
      "\u001b[K\n",
      "Command exit status:\u001b[K\n",
      "  125\u001b[K\n",
      "\u001b[K\n",
      "Command output:\u001b[K\n",
      "  (empty)\u001b[K\n",
      "\u001b[K\n",
      "Command error:\u001b[K\n",
      "  Loading required package: matrixStats\u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘MatrixGenerics’\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\u001b[K\n",
      "      colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\u001b[K\n",
      "      colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\u001b[K\n",
      "      colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\u001b[K\n",
      "      colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\u001b[K\n",
      "      colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\u001b[K\n",
      "      colWeightedMeans, colWeightedMedians, colWeightedSds,\u001b[K\n",
      "      colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\u001b[K\n",
      "      rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\u001b[K\n",
      "      rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\u001b[K\n",
      "      rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\u001b[K\n",
      "      rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\u001b[K\n",
      "      rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\u001b[K\n",
      "      rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\u001b[K\n",
      "      rowWeightedSds, rowWeightedVars\u001b[K\n",
      "  \u001b[K\n",
      "  Loading required package: Biobase\u001b[K\n",
      "  Welcome to Bioconductor\u001b[K\n",
      "  \u001b[K\n",
      "      Vignettes contain introductory material; view with\u001b[K\n",
      "      'browseVignettes()'. To cite Bioconductor, see\u001b[K\n",
      "      'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\u001b[K\n",
      "  \u001b[K\n",
      "  \u001b[K\n",
      "  Attaching package: ‘Biobase’\u001b[K\n",
      "  \u001b[K\n",
      "  The following object is masked from ‘package:MatrixGenerics’:\u001b[K\n",
      "  \u001b[K\n",
      "      rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  The following objects are masked from ‘package:matrixStats’:\u001b[K\n",
      "  \u001b[K\n",
      "      anyMissing, rowMedians\u001b[K\n",
      "  \u001b[K\n",
      "  converting counts to integer mode\u001b[K\n",
      "  estimating size factors\u001b[K\n",
      "  estimating dispersions\u001b[K\n",
      "  gene-wise dispersion estimates: 1 workers\u001b[K\n",
      "  mean-dispersion relationship\u001b[K\n",
      "  final dispersion estimates, fitting model and testing: 1 workers\u001b[K\n",
      "  using 'ashr' for LFC shrinkage. If used in published research, please cite:\u001b[K\n",
      "      Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\u001b[K\n",
      "      https://doi.org/10.1093/biostatistics/kxw041\u001b[K\n",
      "  time=\"2025-10-01T11:42:37+02:00\" level=error msg=\"error waiting for container: unexpected EOF\"\u001b[K\n",
      "\u001b[K\n",
      "Work dir:\u001b[K\n",
      "  /home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_03/work/a4/194845d6f6d9a0fc47f929230de056\u001b[K\n",
      "\u001b[K\n",
      "Container:\u001b[K\n",
      "  quay.io/biocontainers/bioconductor-deseq2:1.34.0--r41hc247a5b_3\u001b[K\n",
      "\u001b[K\n",
      "Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`\u001b[K\n",
      "\u001b[K\n",
      " -- Check '.nextflow.log' file for details\u001b[39m\u001b[K\n"
     ]
    }
   ],
   "source": [
    "!nextflow run nf-core/differentialabundance -r 1.5.0 --input samplesheet_diffabundance.csv --outdir differentialabundance-out --contrasts ./data/contrasts_venn.csv --matrix data/salmon.merged.gene_counts.tsv -profile rnaseq,docker -c nf-ms.config -resume"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Explain all the parameters you set and why you set them in this way. If you used or created additional files as input, explain what they are used for."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!nextflow run nf-core/differentialabundance -r 1.5.0 --input samplesheet_diffabundance.csv --outdir differentialabundance-out --contrasts ./data/contrasts.csv --matrix data/salmon.merged.gene_counts.tsv -profile rnaseq,docker -c nf-ms.config -resume\n",
    "\n",
    "- `--input samplesheet_diffabundance.csv`: Samplesheet containing information about the samples, their conditions and replicates\n",
    "- `--contrasts data/constrasts_venn.csv`: CSV containing information about the contrasts which are used to compare across conditions. Adapted to account for all conditions used in the venn diagram of the paper.\n",
    "- `diff max`: Threshold of 0.5 log2fc f, asame as in the paper.\n",
    "- `--matrix data/salmon.merged.gene_counts.tsv`: Output from rnaseq, containing the gene x sample count matrix.\n",
    "- `-profile rnaseq,docker`: Profile information, assuming rnaseq output\n",
    "- `-c nf-ms.config`: Config file to prevent memory problems\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What were the outputs of the pipeline?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A report which summarizes all individual steps, especially the differential analysis. This includes outlier status, PCA (2D, 3D), a volcano plot and details about all DEGs of both contrasts, as well as parameters used for each individual step."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Would you exclude any samples? If yes, which and why?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As evident per the PCA, Samples `SNI_Sal_2` and `SNI_Sal_4` are outliers in this analysis, and distort the variance of all other sampls significantly, and can be excluded."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many genes were differentially expressed in each contrast? Does this confirm what the paper mentions?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- SNI_oxy vs. SNI_Sal: 1 upregulated, 17 downregulated.\n",
    "- Sham_oxy vs. Sham_Sal: 7 upregulated, 0 downregulated.\n",
    "\n",
    "The paper claims multiple thousands of DEGs, which cannot be confirmed here. The reason for this massive difference in magnitude is likely the different thresholds used: The paper considers a log foldchange of 0.5 to be significant, while a threshold of 1 is used by the pipeline by default. Running the pipeline again with different thresholds might be a more reasonable approach to confirm the findings of the paper."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The paper mentions differentially expressed genes in three brain regions : the NAc, mPFC and VTA. Briefly explain what these 3 regions are."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- NAc: Nucleus accumbens, most significant in motivation, aversion and reward, which related to addiction, which is the primary focus of the paper.\n",
    "- mPFC: Medial prefrontal cortex, an integral processing component of the brain, relevant in memory and attention.\n",
    "- VTA: Ventral tegmental area, is also relevant in the reward mechanism, relating to addiction, and uses dopamine as a prominent neurotransmitter."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Is there anyway from the paper and the material and methods for us to know which genes are included in these regions?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Table 1 might be used for that purpose, although it primarily formated for pathways, instead of brain regions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once you have your list of differentially expressed genes, do you think just communicating those to the biologists would be sufficient? What does the publication state?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Please reproduce the Venn Diagram from Figure 3, not taking into account the brain regions but just the contrasts mentionned."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nf-core",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
