{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 2\n",
    "\n",
    "Today, we will start using nf-core pipelines to find differentially abundant genes in our dataset. \n",
    "We are using data from the following paper: https://www.nature.com/articles/s41593-023-01350-3#Sec10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "1. Please take some time to read through the paper and understand their approach, hypotheses and goals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "### What was the objective of the study?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the results of oxycodone withdrawal on a transcriptomic level on chronic pain afflicted organisms, to find out about the possibility of artificial transcriptomic regulation as a factor to reduce opiod addition and other withdrawal side effects."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "### What do the conditions mean?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "oxy: Treated with oxycodone\n",
    "\n",
    "\n",
    "sal: Treated with saline (control)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What do the genotypes mean?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SNI: prolonged spared nerve injury surgery (chronic pain)\n",
    "\n",
    "Sham: control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imagine you are the bioinformatician in the group who conducted this study. They hand you the raw files and ask you to analyze them.\n",
    "\n",
    "### What would you do?\n",
    "- QC\n",
    "- RNA read mapping\n",
    "- RNA isotype and abundance estimation -> count matrix\n",
    "- Differential analysis\n",
    "- Enrichment analysis (GO/KEGG)\n",
    "\n",
    "### Which groups would you compare to each other? Please also mention which outcome you would expect to see from each comparison.\n",
    "- SNI-Oxy vs SNI-Sal: Addictive effects for SNI-Oxy. Pain-induced immidiate and lasting pertubations for SNI-Saline. \n",
    "- Sham-Oxy vs. Sham-Sal: Addictive effects for SNI-Oxy. No significant effects for Sham-Sal. (Control?)\n",
    "- SNI-Oxy vs. Sham-Oxy: Addictive side effects for both, but lingering pain effects after withdrawal for SNI.\n",
    "- SNI-Sal vs. Sham-Sal: Immidiate and lasting pain effects for SNI-Oxy, no significant effects for Sham-Sal.  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Your group gave you a very suboptimal excel sheet (conditions_runs_oxy_project.xlsx) to get the information you need for each run they uploaded to the SRA.<br>\n",
    "So, instead of directly diving into downloading the data and starting the analysis, you first need to sort the lazy table.<br>\n",
    "Use Python and Pandas to get the table into a more sensible order.<br>\n",
    "Then, perform some overview analysis and plot the results\n",
    "1. How many samples do you have per condition? 8\n",
    "2. How many samples do you have per genotype? 8\n",
    "3. How often do you have each condition per genotype? 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_6565/1699006584.py:4: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df = df.fillna(False)\n",
      "/tmp/ipykernel_6565/1699006584.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df = df.replace(\"x\", True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>RNA-seq</th>\n",
       "      <th>DNA-seq</th>\n",
       "      <th>condition: Sal</th>\n",
       "      <th>Condition: Oxy</th>\n",
       "      <th>Genotype: SNI</th>\n",
       "      <th>Genotype: Sham</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195505</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195506</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195507</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195508</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195509</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195510</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195512</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195513</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195514</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195515</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195517</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195518</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195519</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195520</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Patient  RNA-seq  DNA-seq  condition: Sal  Condition: Oxy  \\\n",
       "Run                                                                     \n",
       "SRR23195505       ?     True    False            True           False   \n",
       "SRR23195506       ?     True    False           False            True   \n",
       "SRR23195507       ?     True    False            True           False   \n",
       "SRR23195508       ?     True    False           False            True   \n",
       "SRR23195509       ?     True    False           False            True   \n",
       "SRR23195510       ?     True    False            True           False   \n",
       "SRR23195511       ?     True    False           False            True   \n",
       "SRR23195512       ?     True    False            True           False   \n",
       "SRR23195513       ?     True    False            True           False   \n",
       "SRR23195514       ?     True    False           False            True   \n",
       "SRR23195515       ?     True    False            True           False   \n",
       "SRR23195516       ?     True    False           False            True   \n",
       "SRR23195517       ?     True    False           False            True   \n",
       "SRR23195518       ?     True    False            True           False   \n",
       "SRR23195519       ?     True    False           False            True   \n",
       "SRR23195520       ?     True    False            True           False   \n",
       "\n",
       "             Genotype: SNI  Genotype: Sham  \n",
       "Run                                         \n",
       "SRR23195505           True           False  \n",
       "SRR23195506          False            True  \n",
       "SRR23195507          False            True  \n",
       "SRR23195508           True           False  \n",
       "SRR23195509           True           False  \n",
       "SRR23195510           True           False  \n",
       "SRR23195511          False            True  \n",
       "SRR23195512          False            True  \n",
       "SRR23195513           True           False  \n",
       "SRR23195514          False            True  \n",
       "SRR23195515          False            True  \n",
       "SRR23195516           True           False  \n",
       "SRR23195517           True           False  \n",
       "SRR23195518           True           False  \n",
       "SRR23195519          False            True  \n",
       "SRR23195520          False            True  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_excel(\"conditions_runs_oxy_project.xlsx\", index_col=\"Run\")\n",
    "df = df.fillna(False)\n",
    "df = df.replace(\"x\", True)\n",
    "df.sort_values(by=[\"condition: Sal\", \"Genotype: SNI\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>RNA-seq</th>\n",
       "      <th>DNA-seq</th>\n",
       "      <th>condition: Sal</th>\n",
       "      <th>Condition: Oxy</th>\n",
       "      <th>Genotype: SNI</th>\n",
       "      <th>Genotype: Sham</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195505</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195506</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195507</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195508</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195509</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195510</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195512</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195513</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195514</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195515</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195517</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195518</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195519</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195520</th>\n",
       "      <td>?</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Patient  RNA-seq  DNA-seq  condition: Sal  Condition: Oxy  \\\n",
       "Run                                                                     \n",
       "SRR23195505       ?     True    False            True           False   \n",
       "SRR23195506       ?     True    False           False            True   \n",
       "SRR23195507       ?     True    False            True           False   \n",
       "SRR23195508       ?     True    False           False            True   \n",
       "SRR23195509       ?     True    False           False            True   \n",
       "SRR23195510       ?     True    False            True           False   \n",
       "SRR23195511       ?     True    False           False            True   \n",
       "SRR23195512       ?     True    False            True           False   \n",
       "SRR23195513       ?     True    False            True           False   \n",
       "SRR23195514       ?     True    False           False            True   \n",
       "SRR23195515       ?     True    False            True           False   \n",
       "SRR23195516       ?     True    False           False            True   \n",
       "SRR23195517       ?     True    False           False            True   \n",
       "SRR23195518       ?     True    False            True           False   \n",
       "SRR23195519       ?     True    False           False            True   \n",
       "SRR23195520       ?     True    False            True           False   \n",
       "\n",
       "             Genotype: SNI  Genotype: Sham Condition Genotype  \n",
       "Run                                                            \n",
       "SRR23195505           True           False       Sal      SNI  \n",
       "SRR23195506          False            True       Oxy     Sham  \n",
       "SRR23195507          False            True       Sal     Sham  \n",
       "SRR23195508           True           False       Oxy      SNI  \n",
       "SRR23195509           True           False       Oxy      SNI  \n",
       "SRR23195510           True           False       Sal      SNI  \n",
       "SRR23195511          False            True       Oxy     Sham  \n",
       "SRR23195512          False            True       Sal     Sham  \n",
       "SRR23195513           True           False       Sal      SNI  \n",
       "SRR23195514          False            True       Oxy     Sham  \n",
       "SRR23195515          False            True       Sal     Sham  \n",
       "SRR23195516           True           False       Oxy      SNI  \n",
       "SRR23195517           True           False       Oxy      SNI  \n",
       "SRR23195518           True           False       Sal      SNI  \n",
       "SRR23195519          False            True       Oxy     Sham  \n",
       "SRR23195520          False            True       Sal     Sham  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alt format: run name, condition, genotype cols\n",
    "df[\"Condition\"] = [\"Sal\" if x else \"Oxy\" for x in df[\"condition: Sal\"]]\n",
    "df[\"Genotype\"] = [\"Sham\" if x else \"SNI\" for x in df[\"Genotype: Sham\"]]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195508</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195509</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195517</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195506</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195514</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195519</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195505</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195510</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195513</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195518</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195507</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195512</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195515</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195520</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Condition Genotype\n",
       "Run                           \n",
       "SRR23195508       Oxy      SNI\n",
       "SRR23195509       Oxy      SNI\n",
       "SRR23195516       Oxy      SNI\n",
       "SRR23195517       Oxy      SNI\n",
       "SRR23195506       Oxy     Sham\n",
       "SRR23195511       Oxy     Sham\n",
       "SRR23195514       Oxy     Sham\n",
       "SRR23195519       Oxy     Sham\n",
       "SRR23195505       Sal      SNI\n",
       "SRR23195510       Sal      SNI\n",
       "SRR23195513       Sal      SNI\n",
       "SRR23195518       Sal      SNI\n",
       "SRR23195507       Sal     Sham\n",
       "SRR23195512       Sal     Sham\n",
       "SRR23195515       Sal     Sham\n",
       "SRR23195520       Sal     Sham"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new = df[[\"Condition\", \"Genotype\"]]\n",
    "df_new = df_new.sort_values(by=[\"Condition\", \"Genotype\"])\n",
    "df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Oxy condition count:  8\n",
      "SNI genotype count:  8\n"
     ]
    }
   ],
   "source": [
    "print(\"Oxy condition count: \", len(df_new[df_new[\"Condition\"] == \"Oxy\"]))\n",
    "print(\"SNI genotype count: \", len(df_new[df_new[\"Genotype\"] == \"SNI\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "They were so kind to also provide you with the information of the number of bases per run, so that you can know how much space the data will take on your Cluster.<br>\n",
    "Add a new column to your fancy table with this information (base_counts.csv) and sort your dataframe according to this information and the condition.\n",
    "\n",
    "Then select the 2 smallest runs from your dataset and download them from SRA (maybe an nf-core pipeline can help here?...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195505</th>\n",
       "      <td>6922564500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195506</th>\n",
       "      <td>7859530800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195507</th>\n",
       "      <td>8063298900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195508</th>\n",
       "      <td>6927786900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195509</th>\n",
       "      <td>7003550100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195510</th>\n",
       "      <td>7377388500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>6456390900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195512</th>\n",
       "      <td>7462857900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195513</th>\n",
       "      <td>8099181600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195514</th>\n",
       "      <td>7226808600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195515</th>\n",
       "      <td>8169101700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>6203117700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195517</th>\n",
       "      <td>6863840400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195518</th>\n",
       "      <td>7908500400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195519</th>\n",
       "      <td>6996050100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195520</th>\n",
       "      <td>7858146000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Bases\n",
       "Run                    \n",
       "SRR23195505  6922564500\n",
       "SRR23195506  7859530800\n",
       "SRR23195507  8063298900\n",
       "SRR23195508  6927786900\n",
       "SRR23195509  7003550100\n",
       "SRR23195510  7377388500\n",
       "SRR23195511  6456390900\n",
       "SRR23195512  7462857900\n",
       "SRR23195513  8099181600\n",
       "SRR23195514  7226808600\n",
       "SRR23195515  8169101700\n",
       "SRR23195516  6203117700\n",
       "SRR23195517  6863840400\n",
       "SRR23195518  7908500400\n",
       "SRR23195519  6996050100\n",
       "SRR23195520  7858146000"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bases = pd.read_csv(\"base_counts.csv\", index_col=\"Run\")\n",
    "bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Bases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6203117700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>6456390900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195517</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6863840400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195505</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6922564500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195508</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6927786900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195519</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>6996050100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195509</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>7003550100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195514</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7226808600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195510</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>7377388500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195512</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7462857900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195520</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7858146000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195506</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7859530800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195518</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>7908500400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195507</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>8063298900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195513</th>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>8099181600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195515</th>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>8169101700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Condition Genotype       Bases\n",
       "Run                                       \n",
       "SRR23195516       Oxy      SNI  6203117700\n",
       "SRR23195511       Oxy     Sham  6456390900\n",
       "SRR23195517       Oxy      SNI  6863840400\n",
       "SRR23195505       Sal      SNI  6922564500\n",
       "SRR23195508       Oxy      SNI  6927786900\n",
       "SRR23195519       Oxy     Sham  6996050100\n",
       "SRR23195509       Oxy      SNI  7003550100\n",
       "SRR23195514       Oxy     Sham  7226808600\n",
       "SRR23195510       Sal      SNI  7377388500\n",
       "SRR23195512       Sal     Sham  7462857900\n",
       "SRR23195520       Sal     Sham  7858146000\n",
       "SRR23195506       Oxy     Sham  7859530800\n",
       "SRR23195518       Sal      SNI  7908500400\n",
       "SRR23195507       Sal     Sham  8063298900\n",
       "SRR23195513       Sal      SNI  8099181600\n",
       "SRR23195515       Sal     Sham  8169101700"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = df_new.merge(bases, on=\"Run\").sort_values([\"Bases\"])\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Bases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6203117700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>6456390900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Condition Genotype       Bases\n",
       "Run                                       \n",
       "SRR23195516       Oxy      SNI  6203117700\n",
       "SRR23195511       Oxy     Sham  6456390900"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "smallest_sras = combined.iloc[0:2]\n",
    "smallest_sras"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m\u001b[38;5;232m\u001b[48;5;43m N E X T F L O W \u001b[0;2m  ~  \u001b[mversion 25.04.7\u001b[m\n",
      "\u001b[K\n",
      "Launching\u001b[35m `https://github.com/nf-core/fetchngs` \u001b[0;2m[\u001b[0;1;36minsane_baekeland\u001b[0;2m] DSL2 - \u001b[36mrevision: \u001b[0;36m8ec2d934f9 [master]\u001b[m\n",
      "\u001b[K\n",
      "\u001b[33mWARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\u001b[39m\u001b[K\n",
      "\n",
      "\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "                                        \u001b[0;32m,--.\u001b[0;30m/\u001b[0;32m,-.\u001b[0m\n",
      "\u001b[0;34m        ___     __   __   __   ___     \u001b[0;32m/,-._.--~'\u001b[0m\n",
      "\u001b[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \u001b[0;33m}  {\u001b[0m\n",
      "\u001b[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \u001b[0;32m\\`-._,-`-,\u001b[0m\n",
      "                                        \u001b[0;32m`._,._,'\u001b[0m\n",
      "\u001b[0;35m  nf-core/fetchngs v1.12.0-g8ec2d93\u001b[0m\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "\u001b[1mCore Nextflow options\u001b[0m\n",
      "  \u001b[0;34mrevision       : \u001b[0;32mmaster\u001b[0m\n",
      "  \u001b[0;34mrunName        : \u001b[0;32minsane_baekeland\u001b[0m\n",
      "  \u001b[0;34mcontainerEngine: \u001b[0;32mdocker\u001b[0m\n",
      "  \u001b[0;34mlaunchDir      : \u001b[0;32m/home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_02\u001b[0m\n",
      "  \u001b[0;34mworkDir        : \u001b[0;32m/home/marcelullose/Documents/SoSe25/ComputationalWorkflows/computational-workflows-2025/notebooks/day_02/work\u001b[0m\n",
      "  \u001b[0;34mprojectDir     : \u001b[0;32m/home/marcelullose/.nextflow/assets/nf-core/fetchngs\u001b[0m\n",
      "  \u001b[0;34muserName       : \u001b[0;32mmarcelullose\u001b[0m\n",
      "  \u001b[0;34mprofile        : \u001b[0;32mdocker\u001b[0m\n",
      "  \u001b[0;34mconfigFiles    : \u001b[0;32m\u001b[0m\n",
      "\n",
      "\u001b[1mInput/output options\u001b[0m\n",
      "  \u001b[0;34minput          : \u001b[0;32mids.csv\u001b[0m\n",
      "  \u001b[0;34moutdir         : \u001b[0;32mfastq/\u001b[0m\n",
      "\n",
      "\u001b[1mMax job request options\u001b[0m\n",
      "  \u001b[0;34mmax_memory     : \u001b[0;32m4GB\u001b[0m\n",
      "\n",
      "!! Only displaying parameters that differ from the pipeline defaults !!\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "If you use nf-core/fetchngs for your analysis please cite:\n",
      "\n",
      "* The pipeline\n",
      "  https://doi.org/10.5281/zenodo.5070524\n",
      "\n",
      "* The nf-core framework\n",
      "  https://doi.org/10.1038/s41587-020-0439-x\n",
      "\n",
      "* Software dependencies\n",
      "  https://github.com/nf-core/fetchngs/blob/master/CITATIONS.md\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_IDS_TO_RUNINFO -\u001b[K\n",
      "\u001b[2A\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_IDS_TO_RUNINFO -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_RUNINFO_TO_FTP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…FETCHNGS:SRA:\u001b[mSRA_FASTQ_FTP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2m\u001b[mNFC…USTOM_SRATOOLSNCBISETTINGS -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…SRATOOLS:\u001b[mSRATOOLS_PREFETCH -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TOOLS:\u001b[mSRATOOLS_FASTERQDUMP -\u001b[K\n",
      "\u001b[7A\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_IDS_TO_RUNINFO -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_RUNINFO_TO_FTP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…FETCHNGS:SRA:\u001b[mSRA_FASTQ_FTP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2m\u001b[mNFC…USTOM_SRATOOLSNCBISETTINGS -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…SRATOOLS:\u001b[mSRATOOLS_PREFETCH -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TOOLS:\u001b[mSRATOOLS_FASTERQDUMP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFCORE_FETCHNGS:SRA:\u001b[mASPERA_CLI -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_TO_SAMPLESHEET -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…RA:\u001b[mMULTIQC_MAPPINGS_CONFIG -\u001b[K\n",
      "\u001b[10A\n",
      "\u001b[2m[\u001b[0;34mfc/3637bb\u001b[0;2m] \u001b[0;2m\u001b[mNFC…S_TO_RUNINFO\u001b[33;2m (\u001b[0;33mSRR23195516\u001b[2m)\u001b[m\u001b[2m |\u001b[m 2 of 2\u001b[2m, cached: 2\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_RUNINFO_TO_FTP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…FETCHNGS:SRA:\u001b[mSRA_FASTQ_FTP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2m\u001b[mNFC…USTOM_SRATOOLSNCBISETTINGS -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…SRATOOLS:\u001b[mSRATOOLS_PREFETCH -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TOOLS:\u001b[mSRATOOLS_FASTERQDUMP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFCORE_FETCHNGS:SRA:\u001b[mASPERA_CLI -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_TO_SAMPLESHEET -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…RA:\u001b[mMULTIQC_MAPPINGS_CONFIG -\u001b[K\n",
      "\u001b[10A\n",
      "\u001b[2m[\u001b[0;34mfc/3637bb\u001b[0;2m] \u001b[0;2m\u001b[mNFC…S_TO_RUNINFO\u001b[33;2m (\u001b[0;33mSRR23195516\u001b[2m)\u001b[m\u001b[2m |\u001b[m 2 of 2\u001b[2m, cached: 2\u001b[0;32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m71/d2106c\u001b[0;2m] \u001b[0;2mNFC…SRA:\u001b[mSRA_RUNINFO_TO_FTP\u001b[33;2m (\u001b[0;33m2\u001b[2m)\u001b[m\u001b[2m |\u001b[m 2 of 2\u001b[2m, cached: 2\u001b[0;32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…FETCHNGS:SRA:\u001b[mSRA_FASTQ_FTP\u001b[2m |\u001b[m 0 of 2\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2m\u001b[mNFC…USTOM_SRATOOLSNCBISETTINGS -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…SRATOOLS:\u001b[mSRATOOLS_PREFETCH -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TOOLS:\u001b[mSRATOOLS_FASTERQDUMP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFCORE_FETCHNGS:SRA:\u001b[mASPERA_CLI -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_TO_SAMPLESHEET -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…RA:\u001b[mMULTIQC_MAPPINGS_CONFIG -\u001b[K\n",
      "\u001b[10A\n",
      "\u001b[2mexecutor >  local (1)\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34mfc/3637bb\u001b[0;2m] \u001b[0;2m\u001b[mNFC…S_TO_RUNINFO\u001b[33;2m (\u001b[0;33mSRR23195516\u001b[2m)\u001b[m\u001b[2m |\u001b[m 2 of 2\u001b[2m, cached: 2\u001b[0;32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34m71/d2106c\u001b[0;2m] \u001b[0;2mNFC…SRA:\u001b[mSRA_RUNINFO_TO_FTP\u001b[33;2m (\u001b[0;33m2\u001b[2m)\u001b[m\u001b[2m |\u001b[m 2 of 2\u001b[2m, cached: 2\u001b[0;32m ✔\u001b[m\u001b[K\n",
      "\u001b[2m[\u001b[0;34me6/932e0d\u001b[0;2m] \u001b[0;2m\u001b[mNFC…\u001b[33;2m (\u001b[0;33mSRX19144486_SRR23195516\u001b[2m)\u001b[m\u001b[2m |\u001b[m 0 of 2\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2m\u001b[mNFC…USTOM_SRATOOLSNCBISETTINGS -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…SRATOOLS:\u001b[mSRATOOLS_PREFETCH -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…TOOLS:\u001b[mSRATOOLS_FASTERQDUMP -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFCORE_FETCHNGS:SRA:\u001b[mASPERA_CLI -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…NGS:SRA:\u001b[mSRA_TO_SAMPLESHEET -\u001b[K\n",
      "\u001b[2m[\u001b[0;34m-        \u001b[0;2m] \u001b[0;2mNFC…RA:\u001b[mMULTIQC_MAPPINGS_CONFIG -\u001b[K\n",
      "\n",
      "-\u001b[0;35m[nf-core/fetchngs]\u001b[0;31m Pipeline completed with errors\u001b[0m-\n",
      "WARN: =============================================================================\n",
      "  Please double-check the samplesheet that has been auto-created by the pipeline.\n",
      "\n",
      "  Public databases don't reliably hold information such as strandedness\n",
      "  information, controls etc\n",
      "\n",
      "  All of the sample metadata obtained from the ENA has been appended\n",
      "  as additional columns to help you manually curate the samplesheet before\n",
      "  running nf-core/other pipelines.\n",
      "===================================================================================\n",
      "WARN: Killing running tasks (1)\n"
     ]
    }
   ],
   "source": [
    "!nextflow run nf-core/fetchngs --max_memory \"4GB\" -profile docker --input ids.csv --outdir fastq/ -resume"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While your files are downloading, get back to the paper and explain how you would try to reproduce the analysis.<br>\n",
    "When you are done with this shout, so we can discuss the different ideas."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nf-core",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
